---
title: "Testis Integration"
output:
  html_notebook: default
  html_document:
    df_print: paged
  pdf_document: default
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 


```{r load_libraries, include = FALSE}
setwd('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/')
directory = '/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/'
#import library
library(plyr)
suppressMessages(library(dplyr))
suppressMessages(library(Seurat))
suppressMessages(library(Matrix))
suppressMessages(library(gplots))
suppressMessages(library(ggplot2))
suppressMessages(library(openxlsx))
suppressMessages(library(cowplot))
suppressMessages(library(patchwork))
suppressMessages(library(repr))
suppressMessages(library(cowplot))
suppressMessages(library("viridis"))
suppressMessages(library('pals'))
suppressMessages(library(RColorBrewer))
suppressMessages(library(wesanderson))
suppressMessages(library("stringr"))
suppressMessages(library("magick"))
suppressMessages(library('pdftools'))
library(enrichR)
library(ggpubr)


options(future.globals.maxSize = 3145728000) 

SO_dir = "/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/Seurat_objects/"
dir.create(SO_dir)

```

LOAD OSR object
```{r load prevous analysis}
#load objects
OBJ_dir = '/Users/tascini.annasofia/Dropbox (HSR Global)/WORKSPACE/Alfano/AlfanoM_904_infertilita_epigenetics/7_bioinfo/SeuratObjects/'

start_time <- Sys.time()

load(paste(OBJ_dir,"Somatic_integration_beforeplots.RData", sep=''))

load(paste(SO_dir, "iNOA_GSE124263_2D_TCLreassigned.integrated", sep = ''))
iNOA_GSE124263_2D_TCLreassigned.integrated = final_int_obj

load(paste(SO_dir, "iNOA_GSE142585_D4.integrated", sep = ''))
iNOA_GSE142585_D4.integrated = final_int_obj

load(paste(SO_dir, "iNOA_Final_GSE124263_GSE112013_iNOA.integrated", sep = ''))
iNOA_Final_GSE124263_GSE112013_iNOA.integrated = final_int_obj

load(paste(SO_dir, "iNOA_Final_GSE124263_GSE112013_iNOA.integrated", sep = ''))
iNOA_Final_GSE124263_GSE112013_iNOA.integrated = final_int_obj

load('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/testis_somatic_OA')
testis_somatic_OA = testis_somatic

end_time <- Sys.time()
end_time - start_time

iNOA_donor1 <- RenameCells(iNOA_donor1, add.cell.id = 'iNOA1_')
iNOA_donor2 <- RenameCells(iNOA_donor2, add.cell.id = 'iNOA2_')
iNOA_donor3 <- RenameCells(iNOA_donor3, add.cell.id = 'iNOA3_')

```

Functions
```{r functions, include = FALSE}
Gene_conversion = function(x, gtf_dictionary, gtf_OSR_dictionary){
  # convert gene name from two different versions
  gene.ensembl =  gtf_dictionary[which(gtf_dictionary$gene_name == x),]$gene_id
  genec = gtf_OSR_dictionary[gtf_OSR_dictionary$gene_id %in% gene.ensembl,]$gene_name
  if (x %in% genec | length(genec) == 0) {genec = x}
  newgene  = paste(genec, collapse = ';')
  return(newgene)
}


myDGE = function(Seurat.object, cell_type, outdir = './') {
  azoospermia.response <- FindMarkers(Seurat.object,
                                      ident.1 = paste(cell_type,"_iNOA", sep=''),
                                      ident.2 = paste(cell_type,"_CTL", sep=''), 
                                      verbose = FALSE)
  write.xlsx(azoospermia.response,
             file= paste(outdir,'DGE_', cell_type,'.iNOA.vs.CTRL.xlsx',sep=''), 
             row.names = T,
             asTable = T)
    return(azoospermia.response)
}

Plot_sign <- function(Seraut.object, signature, operator = sum, title = '') {
    x <- Seraut.object
    DefaultAssay(x) <- "RNA"
    x[["Sign_exp"]] <- apply(FetchData(object = x, 
                                       vars = signature),
                             1,
                             operator)
    FP <- FeaturePlot(x, reduction = "umap", 
                      features = 'Sign_exp', 
                      label = T, 
                      order=T,
                      repel = T,
                      cols = c("lightgrey", "red")) +
                      #cols = as.vector(coolwarm(5))) +
    theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=9, face="italic"),
          axis.text.x = element_text(angle = 90, face = "bold", color = 'black', size=12, hjust =1), 
          axis.title.x = element_text(face = "bold", color = "black", size = 14),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=12),
          axis.title.y = element_text(face = "bold", color = "black", size = 14),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
    labs(title = title, 
         subtitle = paste('MarkerGenes: ',toString(signature), sep=''), 
         x = "UMAP 1", y = "UMAP 2") 
    return(FP)
}




```


Colors
```{r color, include = FALSE}
colors_Alf <- data.frame(celltype = c("SSCs","Differentiated S'gonia",
                                      "Early primary S'cytes",
                                      "Late primary S'cyte",
                                      "Round S'tidis",
                                      "Elong. S'tidis S1",
                                      "Elong. S'tidis S2",
                                      "Sperm",
                                      "Sperm1",
                                      "Sperm2",
                                      "Primary S'cytes",
                                      "LEY",
                                      "MYD",
                                      "SRT",
                                      "MCR",
                                      "TCL",
                                      "END",
                                      "END2",
                                     "STRO",
                                     "UND",
                                     "CTLonly"), 
                        cols = c(brewer.pal(9, 'Oranges')[2:9],
                                 '#932F04',
                                 '#7F2704',
                                 '#FD9E54',
                                 '#E41A1C', #L
                                 '#3399FF', #M
                                 '#FFCC33', #S
                                 '#9900FF', #M
                                 '#FF33CC', #Tcell
                                 '#4DAF4A',
                                 '#4DAF9A',
                                'cadetblue2',
                                '#999999',
                                'grey'))
rownames(colors_Alf) <- colors_Alf$celltype
```



**Dataset GEO GSE112013**

Dataset of adult testis from 3 healthy donors [Guo et al.](https://doi.org/10.1038/s41422-018-0099-2).
Apoptotic tissue, cryopreserved. chromium 10x platform. Already analysed in the previous summition.
```{r GSE112013, include = TRUE}
dataset = "GSE112013" # Guo et al
DimPlot(SO_Guoetal, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="right",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 

# assign somatic cells
testis_somatic <- subset(SO_Guoetal, ident = c('LEY','MYD','MCR','STRO','END', 'SRT')) # all cells
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)


```

**Dataset GEO GSE120506**

Dataset of infant testis from [Guo et al.](https://doi.org/10.1038/s41422-018-0099-2).
Autoptic tissue, cryopreserved. chromium 10x platform.
The re-analysis of this samples does not show division between somatic and spermatogonial cells.
```{r GSE120506, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE120506'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
DimPlot(testis, reduction = "tsne", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "TSNE 1", y = "TSNE 2") 
# germiline markers
FeaturePlot(testis, features = c('DDX4', 'HMGA1'), order = T, reduction = 'tsne') 
# assign somatic cells
testis_somatic <- testis # all cells
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
```

**Dataset GEO GSE142585**

Healthy adult testis from [Shami et al.](10.1016/j.devcel.2020.05.010).
Description of Tcells among somatic populations. autoptic tissue +  cryopreserved in liquid nitrogen following a slow freezing. 
Dropseq platform. Alligned on GRCh37.

```{r GSE142585, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE142585'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 

nPC = 20; res = 1; thresh.use = 0.25; min.pct = 0.25; min.diff.pct = -Inf; test.use="wilcox"
cluster.markers = FindAllMarkers(testis, 
                                 thresh.use = thresh.use, 
                                 test.use = test.use, 
                                 min.pct=min.pct, 
                                 min.diff.pct=min.diff.pct, 
                                 only.pos=TRUE)
FeaturePlot(testis,"CCL5")
DimPlot(testis, label = TRUE)
TCL_Shami = subset(testis, idents = '13')
TCL_Shami
save(TCL_Shami, file = paste(SO_dir, "TCL_Shami", sep=''))

testis$Donor = str_sub(testis$orig.ident,1,6)
table(testis$Donor)
DimPlot(testis, group.by = "Donor")

write.csv(cluster.markers, paste('Seurat_',dataset,'.csv', sep =''))
testis[[]]
DimPlot(testis, split.by = "Donor")
# select somatic 
testis_somatic <- subset(testis, idents = c('9','13','12','11','8', '2'))
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
```

**Dataset GEO GSE109037** 

28 organ donors + 2 apoptotyc tissue divided in 3 batches from [Hermann et al.](https://doi.org/10.1016/j.celrep.2018.10.026). 
chromium 10x platform. Alligned on GRCh38 with cellranger.
```{r GSE109037, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE109037'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
# select somatic 
testis_somatic <- subset(testis, idents = c('0','1','18','19','17'))
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
```

**Dataset GEO GSE124263**
testis from 2 infant and 2 adult from [Sohni et al](https://doi.org/10.1016/j.celrep.2019.01.045) - Miles Wilkinson group.
Adult testicular biopsies were obtained from two fertile men aged 37- and 42-years, undergoing vasectomy.  Cryopreserved using freezing media composed of 10% DMSO + 90% FBS under controlled cooling conditions in a freezing container. 
Infant testes were obtained from unrelated day 2 and day 7 neonates who died as a result of non-testicular-related medical dysfunction; they were cryopreserved in vials in liquid nitrogen and later transported to UCSD on dry ice.
```{r GSE124263, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE124263'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
p1 = DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p2 = DimPlot(testis, reduction = "umap", group.by = 'donor', label = F, order = T, pt.size = 1,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="right",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p1 | p2
# select somatic 
testis_somatic <- subset(testis, idents = c('0','1','3','4','5','8','11','15'))
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
Idents(testis_somatic) <- "donor"
testis_somatic_adult <- subset(testis_somatic, idents = c('adult_Donor1', 'adult_Donor2'))
Idents(testis_somatic_adult) <- "seurat_clusters"
Idents(testis_somatic) <- "seurat_clusters"
table(Idents(testis_somatic_adult))

assign(paste('testis_somatic_adult_',dataset, sep = ''), testis_somatic_adult)

head(testis_somatic_adult[[]])
Idents(testis_somatic_adult) <- "donor"
d1 = subset(testis_somatic_adult, idents = "adult_Donor1")
barcodes_d1 = row.names(d1[[]])
length(barcodes_d1); head(barcodes_d1)
cleanbarcode_d1 = str_sub(barcodes_d1, start = 3, end = -3); head(cleanbarcode_d1)
write.table(x = cleanbarcode_d1, file = 'somatic_GSE124263_donor1_whitelist.txt', sep ='\t', col.names = FALSE, row.names = FALSE, quote = FALSE)

d2 = subset(testis_somatic_adult, idents = "adult_Donor2")
head(d2[[]]); tail(d2[[]])
barcodes_d2 = row.names(d2[[]])
length(barcodes_d2); head(barcodes_d2)
cleanbarcode_d2 = str_sub(barcodes_d2, start = 3, end = -3); head(cleanbarcode_d2)
write.table(x = cleanbarcode_d2, file = 'somatic_GSE124263_donor2_whitelist.txt', sep ='\t', col.names = FALSE, row.names = FALSE, quote = FALSE)
getwd()

# plot RPL MW

Idents(testis) <- "donor"
table(Idents(testis))
testis_adult = subset(testis, idents = c("adult_Donor1", "adult_Donor2"))
Idents(testis_adult) <- "seurat_clusters"

testis_adult <- FindNeighbors(testis_adult, dims = 1:nPC, verbose = FALSE)
testis_adult <- FindClusters(testis_adult, resolution = res, verbose = FALSE)
testis_adult <- RunUMAP(testis_adult, dims = 1:nPC, verbose = FALSE)
DimPlot(testis_adult)

testis_adult[["perc.Ribosomal"]] <- PercentageFeatureSet(testis_adult, pattern = "^RP")
#DimPlot(testis_somatic_adult)
x = testis_adult
new_cluster_id = c("LEY", "Differentiated S'gonia", "MYD", "Early primary S'cytes", "END", "Late primary S'cyte",
                   "SSCs", "MCR", "STRO", "Sperm", "SRT")
names(new_cluster_id) <- levels(x)
x <- RenameIdents(x, new_cluster_id)
RP_lit <- FeaturePlot(x,
                      reduction = "umap", 
                      features = "perc.Ribosomal", 
                      pt.size = 1,
                      order = T,
                      repel = T,
                      label = T, 
                      max.cutoff = 'q98',
                      cols = c("lightgrey", "red")) + ggtitle("GSE124263")
RP_lit
ggsave(paste(outdir, "SI/","GSE124263_RP.png", sep=''), width = 7, height = 6)

```

**Dataset GEO GSE134144** 

4 prepubertal testis (from juvenile males of 7, 11, 13, and 14 years old)  from [Guo et al.](https://doi.org/10.1016/j.stem.2019.12.005).
Aupoptotic tissue - cryopreserved.  chromium 10x platform. Alligned to hg38 with cell ranger.
```{r GSE134144, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE134144'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
p1 = DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
if (dataset == 'GSE134144'){
  barcodes_label = data.frame(barcode = row.names(testis[[]]), 
                              donor = str_sub(row.names(testis[[]]),1,3))
  testis$donor =  barcodes_label$donor
}
p2 = DimPlot(testis, reduction = "umap", group.by = 'donor', label = F, order = T, pt.size = 1,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="right",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p1 | p2
# select somatic 
testis_somatic <- subset(testis, idents = c('0','1','2','3','4','6','7','11'))
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
```
**Dataset GEO GSE134144 2** 

2 transfemale testis (from a 60 and years old adults) from [Guo et al.](https://doi.org/10.1016/j.stem.2019.12.005).
Aupoptotic tissue - cryopreserved.  chromium 10x platform. Alligned to hg38 with cell ranger.

```{r GSE134144_trans, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE134144_t'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
p1 = DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p2 = DimPlot(testis, reduction = "umap", group.by = 'donor', label = F, order = T, pt.size = 1,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="right",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p1 | p2
# select somatic 
testis_somatic <- subset(testis, idents = c('0','1','2','3','4','5','6','9','10'))
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
```

**Dataset GEO GSE125372** 
Two donors with obstructive etiology for infertility from [Xia et al.](https://doi.org/10.1016/j.cell.2019.12.015)
Tissue from testicular sperm extraction (TESE) surgery + freshly processed.
inDrop Single Cell RNA Seq Kit (1CellBio, Cat. 10196) on the inDrop microfluidics system (1CellBio, Cat. 10256-01) + in-house pipeline
```{r GSE125372, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE125372'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
p1 = DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p2 = DimPlot(testis, reduction = "umap", group.by = 'donor', label = F, order = T, pt.size = 1,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="right",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p1 | p2
# select somatic 
testis_somatic <- subset(testis, idents = c('7','9','12','13','14'))
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
testis
testis_somatic
733/3877*100
```


**Dataset GEO GSE106487** 
2 normal, 7 OA (2854 cell in tot), 1 NOA (174 cells) from [Wang et al.](https://doi.org/10.1016/j.stem.2018.08.007).
Smart-seq2 protocol, hg19 human, TopHat, normalized TPM.
```{r GSE106487, include = TRUE}
#load object 
datasets_dir = ('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/')
dataset = 'GSE106487'
load(paste(datasets_dir, 'testis_',dataset, sep = ''))
assign(paste('testis_',dataset, sep = ''), testis)
p1 = DimPlot(testis, reduction = "umap", label = T, order = T, pt.size = 2,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="top",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p2 = DimPlot(testis, reduction = "umap", group.by = 'donor', label = F, order = T, pt.size = 1,) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 24),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
        axis.title.y = element_text(face = "bold", color = "black", size = 24),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position="right",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
p1 | p2
# select somatic 
testis_somatic <- subset(testis, idents = c('5','6','8','10','11'))
assign(paste('testis_somatic_',dataset, sep = ''), testis_somatic)
```

**NOA OSR vs healthy adult testis**
Dataset of adult testis:
	• NOA OSR
	• GSE142585 (Shami et al. [2020]): that also includes T cells. Dropseq platform. Genome: GRCh37 from Ensembl release 75. 1-s2.0-S1534580720303993-main.pdf
	• GSE124263 (Sohni et al [2019]): Miles Wilkinson group, 2 adult testicular biopsies were obtained from two fertile men aged 37- and 42-years, undergoing vasectomy.  Genome: hg19 CellRanger v2.2.0.
	• GSE112013 (Guo et al [2018]): already used.

```{bash download_gtf, include = FALSE}
# Download datasets
# hg19 v75
curl ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz -o Homo_sapiens.GRCh37.75.gtf.g
gunzip Homo_sapiens.GRCh37.75.gtf.gz
# hg38 v97
curl ftp://ftp.ensembl.org/pub/release-97/gtf/homo_sapiens/Homo_sapiens.GRCh38.97.gtf.gz -o Homo_sapiens.GRCh38.97.gtf.gz
gunzip Homo_sapiens.GRCh38.97.gtf.gz

# Cellranger References - 3.0.0 (February 7, 2018)
curl ftp://ftp.ensembl.org/pub/grch37/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh37.87.gtf.gz -o Homo_sapiens.GRCh37.87.gtf.gz
gunzip Homo_sapiens.GRCh37.87.gtf.gz

# Cellranger References - 2.1.0 (February 7, 2018)
curl ftp://ftp.ensembl.org/pub/grch37/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh37.82.gtf.gz -o Homo_sapiens.GRCh37.82.gtf.gz
gunzip Homo_sapiens.GRCh37.82.gtf.gz

# Cellranger References - 3.0.0 (February 7, 2018) - hg38
curl ftp://ftp.ensembl.org/pub/grch37/release-93/gtf/homo_sapiens/Homo_sapiens.GRCh38.93.gtf.gz -o Homo_sapiens.GRCh38.93.gtf.gz
gunzip Homo_sapiens.GRCh38.93.gtf.gz



```


*NOA OSR + GSE142585*
```{r dictionary_datasets: NOA OSR + GSE142585}
############### GSE142585 ###############
dataset = "GSE142585"
counts_file_list = list()
counts_file_list[["GSE142585"]] = "/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/GSE142585_MergedHumanTestis4_CountMatrix.txt"

start_time <- Sys.time()
matrix_sparse = as.sparse(read.table(file = counts_file_list[[dataset]], 
                                     sep="\t", 
                                     header = TRUE, 
                                     quote = "", 
                                     row.names = 1))
end_time <- Sys.time()
end_time - start_time
#

#writeMM(matrix_sparse,file=paste(dataset, '_matrix.mtx', sep=''))
#matrix_sparse = readMM(file=paste(dataset, '_matrix.mtx', sep=''))

d_Genes = row.names(matrix_sparse)


#DefaultAssay(testis_somatic_GSE142585) <- "RNA"
#d_Genes <- row.names(testis_somatic_GSE142585)
#venn::venn(list(GSE142585 = d_Genes , OSR=Tascini_Genes))

gtf <- rtracklayer::import.gff('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/Homo_sapiens.GRCh37.75.gtf')
gtf_df = as.data.frame(gtf)[,c("gene_id", "gene_name")]
gtf_dictionary = gtf_df %>% distinct(gene_id, gene_name, .keep_all = TRUE)

wrong_genes = setdiff(d_Genes, gtf_dictionary$gene_name)
if (length(wrong_genes) != 0) {print("Check dataset genes")} 


gtf_OSR = rtracklayer::import.gff('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/Homo_sapiens.GRCh38.97.gtf')
gtf_OSR_df = as.data.frame(gtf_OSR)[,c("gene_id", "gene_name")]
gtf_OSR_dictionary = gtf_OSR_df %>% distinct(gene_id, gene_name, .keep_all = TRUE)
  
wrong_genes_OSR = setdiff(Tascini_Genes, gtf_OSR_dictionary$gene_name)
if (length(wrong_genes_OSR) != 0) {print("Check OSR genes")} 

```



Convert the gene names from ensmbl75 to ensmbl97 using alias
```{r gene_conversion: NOA OSR + GSE142585}
# d_genes are the genes of the dataset to check
# The gtf_dictionaries are dataframes with the equivalence berween gene_name (symbol) and gene_id (gene.ensembl) 

start_time <- Sys.time()
conv = sapply(d_Genes, Gene_conversion, gtf_dictionary = gtf_dictionary, gtf_OSR_dictionary = gtf_OSR_dictionary)
end_time <- Sys.time()
end_time - start_time

gene = data.frame(old_name = as.character(names(conv)), new_name = as.character(conv), stringsAsFactors = FALSE)
gene[grep(';',conv),]
gene[gene$old_name=="HIST1H4A", "new_name"] = "HIST2H4A"
gene[gene$old_name=="NBPF10", "new_name" ] = "NBPF9"
gene[gene$old_name=="SNORD113","new_name" ] = "RF00181"
gene[gene$old_name=="ZNF223", "new_name"] = "ZNF223"
gene[gene$old_name=="RP11-763B22.7", "new_name"] = "RP11-763B22.7"

p1 = venn::venn(list(GSE142585 = gene$old_name, OSR=Tascini_Genes))
p2 = venn::venn(list(GSE142585 = gene$new_name, OSR=Tascini_Genes))

#matrix_sparse_c = matrix_sparse
#row.names(matrix_sparse_c) = gene$new_name
```

```{r object_preparation}
matrix_sparse_c = matrix_sparse
row.names(matrix_sparse_c) = gene$new_name
testis_GSE142585.raw <- CreateSeuratObject(counts       = matrix_sparse_c, 
                                           project      = dataset, 
                                           min.cells    = 5, 
                                           min.features = 200)
testis_GSE142585.raw

whichcell = row.names(testis_somatic_GSE142585[[]])
testis_GSE142585.raw.somatic = subset(testis_GSE142585.raw, cells = whichcell)
testis_GSE142585.raw.somatic$condition = 'CTL'
testis_GSE142585.raw.somatic$source = 'GSE142585'

testis_GSE142585.raw.somatic <- RenameCells(testis_GSE142585.raw.somatic, add.cell.id = 'GSE142585_')
testis_GSE142585.raw.somatic[[]]
testis_GSE142585.raw.somatic$Donor = str_sub(testis_GSE142585.raw.somatic$orig.ident,1,6)
table(testis_GSE142585.raw.somatic$Donor)
testis_GSE142585.raw.somatic$source = paste(testis_GSE142585.raw.somatic$source, 
                                            testis_GSE142585.raw.somatic$Donor, 
                                            sep = "_")
Idents(testis_GSE142585.raw.somatic) <- "orig.ident"
testis_GSE142585_list = SplitObject(testis_GSE142585.raw.somatic, split.by = "Donor")

```


Integration

```{r integration workflow GSE142585}
source('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/integration_function.R')
start_time <- Sys.time()

dat = 'GSE142585_D4'
dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'_iNOA/', sep ='')

somatic.list <- c(list(iNOA_donor1, iNOA_donor2, iNOA_donor3), testis_GSE142585_list) 
somatic.list
names(somatic.list) <- c('iNOA1', 'iNOA2','iNOA3',
                         "GSE142585_D1","GSE142585_D2","GSE142585_D3", "GSE142585_D4")

int_obj = integration_workflow(somatic.list = somatic.list , 
                               dataset_iNOA_dir = dir, 
                               dataset = dat)
int_obj
DimPlot(int_obj)
table(int_obj$source)
cell_type <- c("MYD",
                "LEY",
                "MCR",
                "STRO",
                "END",
                "TCL",
                "SRT",
                "UND")


final_int_obj = integration_postprocessing(SO = int_obj,
                                           new_cluster_id = cell_type, 
                                         dataset = dat, 
                                           dataset_iNOA_dir = dir,
                                           res = 0.2, 
                                           colors_Alf = colors_Alf)
end_time <- Sys.time()
end_time - start_time
save(final_int_obj, file = paste(SO_dir,'iNOA_',dat,'.integrated', sep =''))
                           
```



**iNOA OSR + GSE124263**

```{r dictionary_datasets: NOA OSR + GSE124263}
############### GSE124263 ###############
dataset = "GSE124263"
counts_file_list = list()
counts_file_list[["GSE124263"]] = '/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/GSE124263_RAW/'

start_time <- Sys.time()
matrix_sparse <- Read10X(data.dir = c(paste(counts_file_list[[dataset]],'neonatal_day2', sep=''),
                                      paste(counts_file_list[[dataset]],'neonatal_day7', sep=''),
                                      paste(counts_file_list[[dataset]],'adult_Donor1', sep=''),
                                      paste(counts_file_list[[dataset]],'adult_Donor2', sep='')),
                         gene.column = 1)

# matrix_sparse <- Read10X(data.dir = c(paste(counts_file_list[[dataset]],'adult_Donor1', sep=''),
#                                      paste(counts_file_list[[dataset]],'adult_Donor2', sep='')),
#                         gene.column = 1)
                         #unique.features = FALSE)
end_time <- Sys.time()
end_time - start_time

d_Genes = row.names(matrix_sparse)

# GRCh37  Cell Ranger software (v2.2.0)
gtf = rtracklayer::import.gff('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/Homo_sapiens.GRCh37.82.gtf')
gtf_df = as.data.frame(gtf)[,c("gene_id", "gene_name")]
#gtf_dictionary = read.delim(file = paste(counts_file_list[["GSE124263"]], 'adult_Donor1/features.tsv', sep=''), header = FALSE)
gtf_dictionary = gtf_df %>% distinct(gene_id, gene_name, .keep_all = TRUE)
d_Genes_name = gtf_dictionary[gtf_dictionary$gene_id %in% d_Genes,]$gene_name
row.names(matrix_sparse) <- d_Genes_name 

wrong_genes = setdiff(d_Genes, gtf_dictionary$gene_id)
if (length(wrong_genes) != 0) {print("Check dataset genes")} 

gtf_OSR = rtracklayer::import.gff('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/Homo_sapiens.GRCh38.97.gtf')
gtf_OSR_df = as.data.frame(gtf_OSR)[,c("gene_id", "gene_name")]
gtf_OSR_dictionary = gtf_OSR_df %>% distinct(gene_id, gene_name, .keep_all = TRUE)
  
wrong_genes_OSR = setdiff(Tascini_Genes, gtf_OSR_dictionary$gene_name)
if (length(wrong_genes_OSR) != 0) {print("Check OSR genes")} 

```


```{r gene_conversion: NOA OSR + GSE124263}
# d_genes are the genes of the dataset to check
# The gtf_dictionaries are dataframes with the equivalence berween gene_name (symbol) and gene_id (gene.ensembl) 

start_time <- Sys.time()
conv = sapply(d_Genes_name, Gene_conversion, gtf_dictionary = gtf_dictionary, gtf_OSR_dictionary = gtf_OSR_dictionary)
end_time <- Sys.time()
end_time - start_time

gene = data.frame(old_name = as.character(names(conv)), new_name = as.character(conv), stringsAsFactors = FALSE)

for (inx in as.numeric(row.names(gene[grep(';',conv),]))) {
  if (length(gtf_OSR_dictionary[gtf_OSR_dictionary$gene_id %in% d_Genes[inx], "gene_name"]) > 0) {
    gene[inx, "new_name"] = gtf_OSR_dictionary[gtf_OSR_dictionary$gene_id %in% d_Genes[inx], "gene_name"]
  } else {
    print(inx)
    gene[inx, "new_name"] = gene[inx, "old_name"]}
}
gene[grep(';',conv),]
  
p1 = venn::venn(list(GSE124263 = gene$old_name, OSR=Tascini_Genes))
p2 = venn::venn(list(GSE124263 = gene$new_name, OSR=Tascini_Genes))

#matrix_sparse_c = matrix_sparse
#row.names(matrix_sparse_c) = gene$new_name
```

```{r object_preparation GSE124263}
matrix_sparse_c = matrix_sparse
row.names(matrix_sparse_c) = gene$new_name
testis_GSE124263.raw <- CreateSeuratObject(counts       = matrix_sparse_c, 
                                           project      = dataset, 
                                           min.cells    = 5, 
                                           min.features = 200)
testis_GSE124263.raw

testis = testis_GSE124263.raw
if (dataset == 'GSE124263'){
  barcodes_label = data.frame(barcode = row.names(testis[[]]), l = str_sub(row.names(testis[[]]),1,2))
  barcodes_label$donor = "neonatal_day2"
  barcodes_label[grep('2_',barcodes_label$l),]$donor = "neonatal_day7"
  barcodes_label[grep('3_',barcodes_label$l),]$donor = "adult_Donor1"
  barcodes_label[grep('4_',barcodes_label$l),]$donor = "adult_Donor2"
  barcodes_label$dev = "adult"
  barcodes_label[grep('neonatal',barcodes_label$donor),]$dev = "neonatal"
  testis$donor = barcodes_label$donor
  testis$dev = barcodes_label$dev
}
testis_GSE124263.raw = testis

whichcell = row.names(testis_somatic_adult_GSE124263[[]])

testis_GSE124263.raw.somatic = subset(testis_GSE124263.raw, cells = whichcell)
testis_GSE124263.raw.somatic$condition = 'CTL'
testis_GSE124263.raw.somatic$GEO = 'GSE124263'

testis_GSE124263.raw.somatic$source <- paste(testis_GSE124263.raw.somatic$GEO, 
                                             testis_GSE124263.raw.somatic$donor, sep = "_")
head(testis_GSE124263.raw.somatic[[]])
testis_GSE124263.raw.somatic <- RenameCells(testis_GSE124263.raw.somatic, add.cell.id = 'GSE124263_')

```

```{r}
# separate two donor in distinct samples
SO = testis_GSE124263.raw.somatic
Idents(SO) <- "donor"
testis_GSE124263.raw.somatic_D1 = subset(SO, idents = "adult_Donor1")
Idents(testis_GSE124263.raw.somatic_D1) <- "orig.ident"
testis_GSE124263.raw.somatic_D2 = subset(SO, idents = "adult_Donor2")
Idents(testis_GSE124263.raw.somatic_D2) <- "orig.ident"
testis_GSE124263.raw.somatic_D1
testis_GSE124263.raw.somatic_D2
testis_GSE124263.raw.somatic
```

```{r}
source('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/integration_function.R')
options(future.globals.maxSize = 3145728000) 

dat = 'GSE124263_2D'

dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'_iNOA/', sep ='')

somatic.list <- list()
somatic.list <- list(iNOA_donor1, iNOA_donor2, iNOA_donor3, testis_GSE124263.raw.somatic_D1, testis_GSE124263.raw.somatic_D2)
names(somatic.list) <- c('iNOA1', 'iNOA2','iNOA3',"GSE124263_D1", "GSE124263_D2")
print("input_samples:")
print(somatic.list)

int_obj = integration_workflow(somatic.list = somatic.list, 
                               dataset_iNOA_dir = dir, 
                               dataset = dat)
int_obj

DimPlot(int_obj, label = T)
levels(int_obj)

cell2rename <- rownames(int_obj@meta.data[int_obj@meta.data$condition == 'CTL' & 
                                                       int_obj@meta.data$seurat_clusters == '6',])

int_obj.mod <- SetIdent(object = int_obj, cells = cell2rename, value = '3')
levels(int_obj.mod)

cell_type <- c( "MCR",
                "LEY",
                "MYD", 
                "END",
                "STRO",
                "SRT",
                "TCL",
                "UND")

library(enrichR)
dat = 'GSE124263_2D_TCLreassigned'
final_int_obj = integration_postprocessing(SO = int_obj.mod,
                                           new_cluster_id = cell_type, 
                                           dataset = dat, 
                                           dataset_iNOA_dir = dir,
                                           res = 0.2, 
                                           colors_Alf = colors_Alf)

save(final_int_obj, file = paste(SO_dir,'iNOA_',dat,'.integrated', sep =''))


DimPlot(final_int_obj, label = T)
levels(final_int_obj )
```




```{r integration workflow GSE124263}
source('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/integration_function.R')
options(future.globals.maxSize = 3145728000) 

dat = 'GSE124263'

dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'_iNOA/', sep ='')

somatic.list <- list()
somatic.list <- list(iNOA_donor1, iNOA_donor2, iNOA_donor3, testis_GSE124263.raw.somatic)
names(somatic.list) <- c('iNOA1', 'iNOA2','iNOA3',"GSE124263")
print("input_samples:")
print(somatic.list)

int_obj = integration_workflow(somatic.list = somatic.list, 
                               dataset_iNOA_dir = dir, 
                               dataset = dat)
int_obj

DimPlot(int_obj, label = T)
levels(int_obj)
DimPlot(final_int_onj, label = T)
levels(final_int_onj )



cell_type <- c( "LEY",
                "MYD", 
                "END",
                "MCR",
                "STRO",
                "SRT",
                "TCL",
                "UND")

library(enrichR)
final_int_obj = integration_postprocessing(SO = int_obj,
                                           new_cluster_id = cell_type, 
                                           dataset = dat, 
                                           dataset_iNOA_dir = dir,
                                           res = 0.2, 
                                           colors_Alf = colors_Alf)
save(final_int_obj, file = paste(SO_dir,'iNOA_',dat,'.integrated', sep =''))

                           
```

```{r UMAP and Ncells}
head(final_int_obj[[]])
ND = 4

DimPlot(final_int_obj, group.by = 'condition', pt.size = 0.2) + scale_color_manual(values = c('orange', 'dodgerblue2'))
ggsave(filename = paste(dir, 'UMAP_bycondition.png', sep=''),
         width = 10, height = 9,
         units = 'cm')
DimPlot(final_int_obj, group.by = 'source', order = TRUE) + scale_color_manual(values = rainbow(ND))
ggsave(filename = paste(dir, 'UMAP_byDONOR.png', sep=''),
         width = 15, height = 9,
         units = 'cm')
DimPlot(final_int_obj, group.by = "celltype", split.by = 'condition')
table(Idents(final_int_obj))

A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$condition),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                                    final_int_obj$condition),margin = 2)
A$N <- A2$Freq
N <- ggplot(data=A, aes(x=Var1, y=Freq*100, fill=Var2)) +
geom_bar(stat="identity",  color="black", position=position_dodge()) + 
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTRL')) +
theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
      axis.text.x = element_text(angle = 0, face = "bold", color = "black", size=22, hjust =.5), 
      axis.title.x = element_text(face = "bold", color = "black", size = 24),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
      axis.title.y = element_text(face = "bold", color = "black", size = 24),
      legend.title = element_text(size = 0),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position = c(0.9, 0.8),
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(x = "CellType", y = "%")

pdf(paste(dir,'N_iNOAvsCTL.pdf',sep=''),width=10, height=8)
print(N)
dev.off()


A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$source),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                                    final_int_obj$source),margin = 2)
A$N <- A2$Freq
N <- ggplot(data=A, aes(x=Var1, y=Freq*100, fill=Var2)) +
geom_bar(stat="identity",  color="black", position=position_dodge()) + 
scale_fill_manual(values = rainbow(ND))+
theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
      axis.text.x = element_text(angle = 0, face = "bold", color = "black", size=22, hjust =.5), 
      axis.title.x = element_text(face = "bold", color = "black", size = 24),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
      axis.title.y = element_text(face = "bold", color = "black", size = 24),
      legend.title = element_text(size = 0),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position = c(0.8, 0.8),
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(x = "CellType", y = "%")

pdf(paste(dir,'N_differntSources.pdf',sep=''),width=10, height=8)
print(N)
dev.off()


Idents(final_int_obj) <- "celltype"

tail(final_int_obj[[]])
A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$condition),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                                    final_int_obj$condition),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj),]$cols)
N <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
geom_bar(stat="identity",  color="black", position="fill") + 
scale_fill_manual(values = col) +
theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=12, hjust =.5), 
      axis.title.x = element_text(face = "bold", color = "black", size = 14),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
      axis.title.y = element_text(face = "bold", color = "black", size = 14),
      legend.title = element_text(size = 0),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position = "left",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(x = "", y = "fraction of cells")

A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$source),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                                    final_int_obj$source),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj),]$cols)
N2 <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
geom_bar(stat="identity",  color="black", position="fill") + 
scale_fill_manual(values = col) +
#scale_x_discrete(labels=c('GSE124263_adult_Donor1'='GSE124263_#1', 
 #                         'GSE124263_adult_Donor2' = 'GSE124263_#2', 
#                          'iNOA donor 1' = 'iNOA#1', 
 #                         'iNOA donor 2' ='iNOA#2', 
  #                        'iNOA donor 3' ='iNOA#3')) +
theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=8, hjust = .5), 
      axis.title.x = element_text(face = "bold", color = "black", size = 14),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
      axis.title.y = element_text(face = "bold", color = "black", size = 14),
      legend.title = element_text(size = 0),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position = "left",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(x = "", y = "fraction of cells")
N + N2 + plot_layout(guides = "collect")

pdf(paste(dir,'N_iNOAvsCTL.pdf',sep=''),width=15, height=6)
print(N + N2 + plot_layout(guides = "collect"))
dev.off()

table(final_int_obj$source)
```

```{r dge genes}
cell_type <- c( "LEY",
                "MYD", 
                "END",
                "MCR",
                "STRO",
                "SRT")

#cell_type <- c( "LEY",
#               "MYD", 
#                "SRT",
#                "UND")

#COL = row.names(Seurat.object)[grep("^COL",row.names(Seurat.object))]
gene = c("DLK1", "NOTCH2", "IGF1", "IGF2", "INSL3", "HSD17B3",
         "CDKN2A",
         "COL1A1", "COL1A2", "COL4A1", "COL4A2", 
         'H19', 'MEG3', 'MEG8', 'HYMAI', 'PEG3', 'ERAP2', 'RTL1', "KCNQ1OT1")

Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <-"RNA"
Idents(Seurat.object) <- "celltype.cond"

azoospermia.modulated.gene = data.frame()
for (ct in cell_type) {
  #print(ct)
  azoospermia.response.gene <- data.frame()
  azoospermia.response.gene <- FindMarkers(Seurat.object,
                                        ident.1 = paste(ct,"_iNOA", sep=''),
                                        ident.2 = paste(ct,"_CTL", sep=''), 
                                        verbose = FALSE,
                                        feature=gene,
                                        logfc.threshold=0,
                                        min.pct = 0)
    
    azoospermia.response.gene          <- azoospermia.response.gene[gene,] 
    azoospermia.response.gene$geneID   <- rownames(azoospermia.response.gene)
    azoospermia.response.gene$cellType <- ct 
    azoospermia.modulated.gene         <- rbind(azoospermia.modulated.gene, azoospermia.response.gene)  
}

#azoospermia.modulated.gene 
myorder=c('cellType','geneID','avg_logFC','p_val_adj','p_val','pct.1','pct.2')
azoospermia.modulated.gene <- azoospermia.modulated.gene[,myorder]
write.xlsx(azoospermia.modulated.gene,
               file= paste(dir,'azoospermia.modulated.gene.azoospermia.xlsx',sep=''), 
               row.names = FALSE,
               asTable = TRUE)

```


```{r}
##### CDKN2A #####

#CDKN2A
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
Seurat.object_s <- Seurat.object # subset(Seurat.object,  idents = c("LEY", "MYD","SRT"))
#Seurat.object_s = Seurat.object
vp_D <- VlnPlot(Seurat.object_s,
        feature = "CDKN2A",
        #slot = "counts", 
        #log = TRUE,
        split.plot = T,
        pt.size = 2,
       split.by = 'condition') +
ylab("Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTL')) +
scale_color_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTL'))

fp_D <- FeaturePlot(Seurat.object, 
            reduction = "umap", 
            features = c("CDKN2A"), 
            label = T, 
            order= T,
            pt.size = 2,
            max.cutoff = 'q99',
            cols = c("lightgrey", "red"),
            split.by='condition') + theme(legend.position = 'right')

vp_D
fp_D
ggsave(filename = paste(dir,"Featureplot_CDKN2A.png",sep=''),
         plot = fp_D,
         width = 20, height = 10,
         units = 'cm')
ggsave(filename = paste(dir,"Violinplot_CDKN2A.png",sep=''),
         plot = vp_D,
         width = 32, height = 16,
         units = 'cm')
```


```{r DLK1 NOTCH2}
##### DLK1 NOTCH2 #####

#DLK1
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
Idents(Seurat.object) <- "celltype"
Seurat.object_s <- subset(Seurat.object,  idents = c("LEY", "MYD","SRT"))
vp_D <- VlnPlot(Seurat.object_s,
        feature = 'DLK1',
        split.plot = T,
        pt.size = 0,
       split.by = 'condition') +
ylab("Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'))

fp_D <- FeaturePlot(Seurat.object, 
            reduction = "umap", 
            features = c('DLK1'), 
            label = T, 
            order= T,
            max.cutoff = 'q99',
            cols = c("lightgrey", "red"),
            split.by='condition') + theme(legend.position = 'right')


vp_N <- VlnPlot(Seurat.object_s,
        feature = 'NOTCH2',
        split.plot = T,
        pt.size = 0,
       split.by = 'condition') +
ylab("Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTL'))

Seurat.object$condition = plyr::revalue(Seurat.object$condition, c("iNOA"="iGCA", "CTL"="CTL"))

fp_N <- FeaturePlot(Seurat.object, 
            reduction = "umap", 
            pt.size = 1,
            features = c('NOTCH2'), 
            label = T, 
            order= T,
            max.cutoff = 'q99',
            cols = c("lightgrey", "red"),
            split.by='condition') + theme(legend.position = 'right')


fp_blend <- FeaturePlot(Seurat.object, 
            reduction = "umap", 
            pt.size = 1,
            features = c('DLK1','NOTCH2'),
            label = T, 
            order= T,
             blend = T,       
            max.cutoff = 'q99',
            cols = c("lightgrey", "red","green"),
            split.by='condition') + theme(legend.position = 'right')


layout <- '
AACCCC
BBCCCC
' 


pdf(paste(dir,'Figure3_DLK1_NOTCH2_blend.pdf',sep=''),width=19, height=7)
wrap_plots(A = vp_D, B = vp_N, C = fp_blend, design = layout)
dev.off()

im = image_read_pdf(paste(dir,"Figure3_DLK1_NOTCH2_blend.pdf",sep=''),density = 140)
image_write(im, path = paste(dir,"Figure3_DLK1_NOTCH2_blend.tiff", sep=''), format = "tiff")

```

```{r IGF1 IGF2}
######## IGF1 IGF2##########
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
Seurat.object_s <- subset(Seurat.object, idents = c("LEY", "MYD","SRT","END","STRO"))
#Seurat.object_s <- subset(Seurat.object, idents = c("LEY", "MYD"))

vp_IG1 <- VlnPlot(Seurat.object_s,
        feature = c('IGF1'),
        #slot = "counts", 
        #log = TRUE,
        split.plot = T,
        pt.size = 0,
       split.by = 'condition') +
ylab("Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTL'))

vp_IG2 <- VlnPlot(Seurat.object_s,
        feature = c('IGF2'),
        #slot = "counts", 
        #log = TRUE,
        split.plot = T,
        pt.size = 0,
       split.by = 'condition') +
ylab("Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTL'))

Seurat.object$condition = plyr::revalue(Seurat.object$condition, c("iNOA"="iGCA", "CTL"="CTL"))
fp_bl <- FeaturePlot(Seurat.object, 
            reduction = "umap", 
            pt.size = 1,
            features = c('IGF1','IGF2'), 
            label = T, 
            order= T,
             blend = T,       
            max.cutoff = 'q99',
            #cols = c("lightgrey", "red"),
            split.by='condition') + theme(legend.position = 'right')

fp_bl
vp_IG1

layout <- '
AACCCC
BBCCCC
' 
pdf(paste(dir,'Figure3_IGFs_blend.pdf',sep=''),width=19, height=7)
wrap_plots(A = vp_IG1, B = vp_IG2, C = fp_bl, design = layout)
dev.off()

im = image_read_pdf(paste(dir,"Figure3_IGFs_blend.pdf",sep=''),density = 140)
image_write(im, path = paste(dir,"Figure3_IGFs_blend.tiff", sep=''), format = "tiff")


```

```{r COL1A1 COL1A2}
#### COL1A1 COL1A2 ######
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
Seurat.object_s <- subset(Seurat.object,  idents = c("LEY", "MYD","STRO"))
#Seurat.object_s <- subset(Seurat.object,  idents = c("LEY", "MYD"))

vp_IG1 <- VlnPlot(Seurat.object_s,
        feature = c('COL1A1'),
        #slot = "counts", 
        #log = TRUE,
        split.plot = T,
        pt.size = 0,
       split.by = 'condition') +
ylab("Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTL'))

vp_IG2 <- VlnPlot(Seurat.object_s,
        feature = c('COL1A2'),
        #slot = "counts", 
        #log = TRUE,
        split.plot = T,
        pt.size = 0,
       split.by = 'condition') +
ylab("Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','CTL'))

Seurat.object$condition = plyr::revalue(Seurat.object$condition, c("iNOA"="iGCA", "CTL"="CTL"))
fp_bl <- FeaturePlot(Seurat.object, 
            reduction = "umap", 
            pt.size = 1,
            features = c('COL1A1','COL1A2'), 
            label = T, 
            order= T,
             blend = T,       
            max.cutoff = 'q99',
            #cols = c("lightgrey", "red"),
            split.by='condition') + theme(legend.position = 'right')


layout <- '
AACCCC
BBCCCC
' 
pdf(paste(dir,'Figure5_COL1s_blend.pdf',sep=''),width=19, height=7)
wrap_plots(A = vp_IG1, B = vp_IG2, C = fp_bl, design = layout)
dev.off()

im = image_read_pdf(paste(dir,"Figure5_COL1s_blend.pdf",sep=''),density = 140)
image_write(im, path = paste(dir,"Figure5_COL1s_blend.tiff", sep=''), format = "tiff")

   
```


```{r imprinted genes}
relevant_imprinted = c('H19', 'MEG3', 'MEG8','IGF2', 'DLK1', 'HYMAI', 'PEG3', 'ERAP2', 'RTL1', "KCNQ1OT1")
dir.create(paste(dir,'ImprintedGenes/', sep = ''))
dir.create(paste(dir,'ImprintedGenes/VlnPlot_r', sep = ''))
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
so = subset(Seurat.object, idents = c('LEY', 'LEY2', 'MYD'))

for (gene in relevant_imprinted) {
  so[[gene]] <- FetchData(object = so, vars = gene)
  df= so[[]]
  df$condition = plyr::revalue(df$condition, c("iNOA"="iGCA", "CTL"="CTL"))
  v <- ggviolin(df, x = "condition", y = gene, 
                fill = "condition", color = "condition", width = 1,
                add = c('jitter'), add.params = list(size =.5, shape = 1),
                #              add.params = list(size = .5, shape = 20),
                alpha = 0.6,
                facet.by = 'celltype', short.panel.labs = TRUE,
                palette = c('orange', 'dodgerblue2'), 
                error.plot = "crossbar", draw_quantiles = c(0.5), trim = T,
                title = gene, 
                panel.labs.background = list(color = "white", fill = "white", size = 0.5),
                panel.labs.font = list(color = "black", face = "bold", size = 16))
  df$condition
  v = ggpar(v, ylab = 'Expression level', xlab = '')
  v = v + stat_compare_means(aes(group = condition),label = "p.signif", size =  8, 
                             label.x.npc = "center", label.y.npc = 0.85) +
    
    theme(plot.title = element_text(color="black", size=22, face="bold.italic", hjust = 0.5),
          plot.title.position =  'panel',
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 14, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 14),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=14),
          axis.title.y = element_text(face = "bold", color = "black", size = 16),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = 'right',
          panel.spacing = unit('-0.1', "lines"),
          #legend.text = c('',''),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid"))
  assign(paste('v','c',gene,sep = '_'),v)
  pdf(paste(dir, 'ImprintedGenes/VlnPlot_r/','vlnPlot_',gene,'.pdf', sep=''), width = 5, height = 4.5)
  print(v)
  dev.off()
  
}

pdf(paste(dir, 'ImprintedGenes/VlnPlot_r/','vlnPlot_maternal.pdf', sep=''), width = 12, height = 4)
print(v_c_H19 + v_c_MEG3 + v_c_MEG8 + plot_layout(guides = "collect", ncol = 3) & 
        theme(legend.position = "right", legend.direction = 'vertical'))
dev.off()

pdf(paste(dir, 'ImprintedGenes/VlnPlot_r/','vlnPlot_paternal.pdf', sep=''), width = 28, height = 4)
print(v_c_IGF2 + v_c_DLK1 + v_c_HYMAI + v_c_PEG3 + v_c_ERAP2 + v_c_RTL1 + v_c_KCNQ1OT1 + plot_layout(guides = "collect", ncol = 7) & 
        theme(legend.position = "right", legend.direction = 'vertical'))
dev.off()


```

```{r}
######## HSD17B3 ########
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
Seurat.object_s <- subset(Seurat.object,  idents = c("LEY"))

vp_HD <- VlnPlot(Seurat.object_s,
        feature = c('HSD17B3'),
        slot = "counts", 
        log = TRUE,
        split.plot = T,
        pt.size = 1,
       split.by = 'condition') +
ylab("Log Expression Level") +
xlab("") +
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.65, 0.9),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iNOA','CTL'))

Seurat.object$condition = plyr::revalue(Seurat.object$condition, c("iNOA"="iGCA", "CTL"="CTL"))
fp_bl <- FeaturePlot(Seurat.object, 
            reduction = "umap", 
            pt.size = 1,
            features = c('HSD17B3'), 
            label = T, 
            order= T,
            blend = F,       
            max.cutoff = 'q99',
            cols = c("lightgrey", "red"),
            split.by='condition') + theme(legend.position = 'right')



layout <- '
ACCCC
ACCCC
' 

pdf(paste(dir,'Figure_HSD17B3.pdf',sep=''),width=14, height=4)
wrap_plots(A = vp_HD, C = fp_bl, design = layout)
dev.off()

pdf(paste(dir,'Figure_HSD17B3.pdf',sep=''),width=8, height=4)
plot(fp_bl)
dev.off()

im = image_read_pdf(paste(dir,'Figure_HSD17B3.pdf',sep=''),density = 140)
image_write(im, path = paste(dir,'Figure_HSD17B3.tiff', sep=''), format = "tiff")
```
```{r}
################ INSL3 #################
final_int_obj = iNOA_Final_GSE124263_GSE112013_iNOA.integrated
Seraut.object= final_int_obj

CELL_INSL3 <- WhichCells(Seraut.object, slot = 'counts', expression = INSL3 > 1)
somatic.integrated.new.INSL3 <- subset(Seraut.object, cells = CELL_INSL3)


legendiNOA='iGCA'
legendCTRL='CTL'

Seraut.object = somatic.integrated.new.INSL3
azoospermia.response.gene <- FindMarkers(Seurat.object,
                                           ident.1 = paste(ct,"_",c1, sep=''),
                                           ident.2 = paste(ct,"_",c2, sep=''), 
                                           verbose = FALSE,
                                           feature=gene,
                                           logfc.threshold=0,
                                           min.pct = 0)

Idents(Seraut.object) <- "celltype"

cell_type.2.use = 'LEY'
cell_type = c(paste(cell_type.2.use,'_iNOA',sep=''),
              paste(cell_type.2.use,'_CTL',sep=''))
#cell_type = c(paste(cell_type.2.use,'_iNOA',sep=''),
#              paste(cell_type.2.use,'_OA',sep=''))
features = c('INSL3')

    
    Cell.subset <- subset(Seraut.object, idents =  cell_type.2.use, invert=F)
    #subset.counts <- Cell.subset@assays$RNA@counts
    subset.counts <- Cell.subset@assays$RNA@data
    gene.counts.dataframe = data.frame(gene = character(),
                                       cell_barcode = character(),
                                       cell_type = character(),
                                       counts = integer(),
                                       stringsAsFactors=FALSE)
    for (gene in features) {
        gene.counts.tmp = subset.counts[gene,]
        gene.counts.dataframe.entry = data.frame(gene = rep(gene, length(gene.counts.tmp)),
                                                 cell_barcode =  names(gene.counts.tmp),
                                                 condition = Seraut.object[[]][names(gene.counts.tmp),
                                                                               'celltype.cond'],
                                                 counts = (gene.counts.tmp))
        gene.counts.dataframe <- rbind(gene.counts.dataframe, gene.counts.dataframe.entry)
        }
gene.counts.dataframe$condition <- factor(x = gene.counts.dataframe$condition, 
                                          levels = c(paste(cell_type.2.use,"_iNOA",sep=''),paste(cell_type.2.use,"_CTL",sep='')))
                                                     #paste(cell_type.2.use,"_CTL",sep='')))

CVP = ggplot(gene.counts.dataframe, aes(x=gene, y = counts, fill=condition)) +
            geom_split_violin() + 
            #ylim(0,6) +
 #scale_y_log10() +
        scale_x_discrete('LEY') +
            #scale_y_continuous(trans='log') +
            scale_fill_manual(values = c('purple', 'orange')) +
            #geom_jitter(shape=16, size = 0.3, position=position_jitter(0.4)) +  
theme(plot.title = element_text(color="black", size=22, face="bold.italic", hjust = 0.5),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
      axis.text.x = element_blank(),
#          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5),
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.87),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#+
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c(legendiNOA,legendCTRL)) +
labs(title = cell_type.2.use , 
     x = "Genes", 
     y = "Expression Level") 
CVP = CVP +labs(title= 'INSL3',
          x = "LEY", 
          y = "Expression Level") 
assign('INSL3_INSL3cell',CVP)

############
Seraut.object= final_int_obj
Idents(Seraut.object) <- "celltype"

cell_type.2.use = 'LEY'
cell_type = c(paste(cell_type.2.use,'_iNOA',sep=''),
              paste(cell_type.2.use,'_CTL',sep=''))
features = c('INSL3')

    
    Cell.subset <- subset(Seraut.object, idents =  cell_type.2.use, invert=F)
    #subset.counts <- Cell.subset@assays$RNA@counts
    subset.counts <- Cell.subset@assays$RNA@data
    gene.counts.dataframe = data.frame(gene = character(),
                                       cell_barcode = character(),
                                       cell_type = character(),
                                       counts = integer(),
                                       stringsAsFactors=FALSE)
    for (gene in features) {
        gene.counts.tmp = subset.counts[gene,]
        gene.counts.dataframe.entry = data.frame(gene = rep(gene, length(gene.counts.tmp)),
                                                 cell_barcode =  names(gene.counts.tmp),
                                                 condition = Seraut.object[[]][names(gene.counts.tmp),
                                                                               'celltype.cond'],
                                                 counts = (gene.counts.tmp))
        gene.counts.dataframe <- rbind(gene.counts.dataframe, gene.counts.dataframe.entry)
        }
gene.counts.dataframe$condition <- factor(x = gene.counts.dataframe$condition, 
                                          levels = c(paste(cell_type.2.use,"_iNOA",sep=''),paste(cell_type.2.use,"_CTL",sep='')))
                                                     #paste(cell_type.2.use,"_CTL",sep='')))

CVP = ggplot(gene.counts.dataframe, aes(x=gene, y = counts, fill=condition)) +
            geom_split_violin() + 
            #ylim(0,6) +
 #scale_y_log10() +
        scale_x_discrete('LEY') +
            #scale_y_continuous(trans='log') +
            scale_fill_manual(values = c('purple', 'orange')) +
            #geom_jitter(shape=16, size = 0.3, position=position_jitter(0.4)) +  
theme(plot.title = element_text(color="black", size=22, face="bold.italic", hjust = 0.5),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
      axis.text.x = element_blank(),
#          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5),
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = c(0.85, 0.87),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
#+
#geom_boxplot(width = 0.1, outlier.size=1) +
scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c(legendiNOA,legendCTRL)) +
labs(title = cell_type.2.use , 
     x = "Genes", 
     y = "Expression Level") 
CVP = CVP +labs(title= 'INSL3',
          x = "LEY", 
          y = "Expression Level") 
CVP

assign('INSL3_allcell',CVP) 

INSL3_allcell + plot_spacer() + INSL3_INSL3cell
p1 =INSL3_allcell; p2 =INSL3_INSL3cell;
(p1 + theme(plot.margin = unit(c(0,30,0,0), "pt"))) +
(p2 + theme(plot.margin = unit(c(0,0,0,30), "pt")))
pdf(paste(dir,'Figure_INSL3.pdf',sep=''),width=8.5, height=3.5)
plot((p1 + theme(plot.margin = unit(c(0,40,0,15), "pt"))) +
       (p2 + theme(plot.margin = unit(c(0,15,0,40), "pt"))))
dev.off()

im = image_read_pdf(paste(dir,'Figure_INSL3.pdf',sep=''),density = 140)
image_write(im, path = paste(dir,'Figure_INSL3.tiff', sep=''), format = "tiff")



```


```{r pathways UP}
#### Pathway UP heatmaps ####
Seurat.object = final_int_obj
Seurat.object$cell_type = Seurat.object$celltype
Idents(Seurat.object) <- "celltype.cond"
DefaultAssay(Seurat.object) <- "RNA"

dir

CND = 'up'
database = c('Reactome_2016','GO_Biological_Process_2018')
indir =  paste(dir,'enrichR/', sep = '')
indir
cell.type_2_plot = c("LEY", "MYD", "MCR")

pathway_to_plot <- list()
pathway_to_plot[["LEY"]] <- c("Cholesterol biosynthesis Homo sapiens R-HSA-191273",
                              "Extracellular matrix organization Homo sapiens R-HSA-1474244",
                              "Collagen formation Homo sapiens R-HSA-1474290",
                              "Collagen biosynthesis and modifying enzymes Homo sapiens R-HSA-1650814",
                              "collagen fibril organization (GO:0030199)",
                              "IRE1alpha activates chaperones Homo sapiens R-HSA-381070",
                              "regulation of cell migration (GO:0030334)",
                              "regulation of keratinocyte apoptotic process (GO:1902172)")

pathway_to_plot[["MYD"]] <- c("Cholesterol biosynthesis Homo sapiens R-HSA-191273",
                              "type I interferon signaling pathway (GO:0060337)",
                              "regulation of interferon-gamma-mediated signaling pathway (GO:0060334)",
                              "Extracellular matrix organization Homo sapiens R-HSA-1474244",
                              "Collagen biosynthesis and modifying enzymes Homo sapiens R-HSA-1650814",
                              "Scavenging by Class A Receptors Homo sapiens R-HSA-3000480",
                              "regulation of cell differentiation (GO:0045595)",
                              "glutathione derivative metabolic process (GO:1901685)",
                              "negative regulation of cellular macromolecule biosynthetic process (GO:2000113)",
                              "regulation of cell migration (GO:0030334)")


pathway_to_plot[["MCR"]] <- c("MHC class II antigen presentation Homo sapiens R-HSA-2132295",
                              "PD-1 signaling Homo sapiens R-HSA-389948",
                              "Interferon gamma signaling Homo sapiens R-HSA-877300",
                              "regulated exocytosis (GO:0045055)",
                              "regulation of amyloid-beta formation (GO:1902003)",
                              "neutrophil mediated immunity (GO:0002446)",
                              "homotypic cell-cell adhesion (GO:0034109)",
                             "cytokine-mediated signaling pathway (GO:0019221",
                             "production of miRNAs involved in gene silencing by miRNA (GO:0035196)",
                             "regulation of transcription from RNA polymerase II promoter (GO:0006357)",
                             "regulation of protein phosphorylation (GO:0001932)")

outdir = paste(dir, 'Pathways/', sep='') 
dir.create(outdir)

for (i in cell.type_2_plot) {
  print(i)
  ds = Seurat.object
  gl = 6
  if (i == 'MCR') {gl = 3}
    HP.col <- DGE_Heatmap(Seurat.object = ds, 
                             cell_type = i, 
                             enrichR.path = indir,
                             CND = 'up', 
                             database = c('Reactome_2016','GO_Biological_Process_2018'), 
                             pathway_to_plot[[i]],
                          gene.label = gl)
    
    pdf(paste(outdir,'HM_DS_',i,'.pdf',sep=''),  width=9.5, height=9)
    print(HP.col)
    dev.off()
}

######## Dotplots ##########à

pathways2plot = pathway_to_plot
pathways.dataframe = data.frame(cellType = character(),
                                Pathway = character(),
                                gene.ratio = numeric(),
                                p.value = numeric(),
                                p.value.adj = numeric(),
                                stringsAsFactors=FALSE)

all.cell.types= cell.type_2_plot

fx <- function(x) eval(parse(text=enrichR.table$Overlap[x]))
fx <- function(x) eval(parse(text=enrichR.table[x,]$Overlap))

for (cell_type in all.cell.types) {
    print(cell_type)
    enrichR.file = paste(indir,"enrichR_DGE_",cell_type,'_',CND,'.xlsx',sep='')
    enrichR.table = data.frame()
    for (dat in database) {
        Table <- read.xlsx(xlsxFile = enrichR.file, 
                            sheet = dat, 
                            startRow = 1, 
                            colNames = TRUE,
                            rowNames = TRUE, 
                            detectDates = FALSE, 
                            skipEmptyRows = TRUE,
                            skipEmptyCols = TRUE,
                            na.strings = "NA", 
                            fillMergedCells = FALSE)
        
        enrichR.table = rbind(enrichR.table , Table)
    }
    enrichR.table <- enrichR.table[order(enrichR.table$Adjusted.P.value, decreasing = F),]
    N = length(pathways2plot[[cell_type]])
    pathways.dataframe.entry = data.frame(cellType = rep(cell_type, N),
                                          Pathway = pathways2plot[[cell_type]],
                                          gene.ratio = sapply(pathways2plot[[cell_type]], fx),
                                          p.value = enrichR.table[pathways2plot[[cell_type]],]$P.value,
                                          p.value.adj = enrichR.table[pathways2plot[[cell_type]],]$Adjusted.P.value)
    
    #pathways.dataframe <- rbind(pathways.dataframe, pathways.dataframe.entry)
    pathways.dataframe <- pathways.dataframe.entry[order(pathways.dataframe.entry$p.value, decreasing = F),]
    pathways.dataframe$Pathway.num = dim(pathways.dataframe)[1]:1
    pathways.dataframe$Pathway.num = as.factor(pathways.dataframe$Pathway.num )
    removeHomo <- function(str, stopwords='Homo') {
      x <- unlist(strsplit(as.character(str), " "))
      paste(x[!x %in% stopwords], collapse = " ")
    }
    removeSapiens <- function(str, stopwords='sapiens') {
      x <- unlist(strsplit(as.character(str), " "))
      paste(x[!x %in% stopwords], collapse = " ")
    }
    pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeHomo)
    pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeSapiens)
    pathways.dataframe$Pathway <- str_pad(pathways.dataframe$Pathway, width = 70, side = "right", pad = " ")
    
    changeName = TRUE
    if (changeName) {
        if (cell_type == 'MYD') {
            pathways.dataframe$Pathway[5] = "regulation of interferon-gamma-mediated signaling pathway \n (GO:0060334)"
            pathways.dataframe$Pathway[4] = "negative regulation of cellular macromolecule biosynthetic\n process (GO:2000113)"
        }
        }
    
    PP = ggplot(pathways.dataframe, aes(Pathway.num,-log10(p.value))) + 
      geom_point(aes(size = gene.ratio), color = colors_Alf[cell_type,]$cols) +
      scale_size_continuous(range = c(5,15), name = "Gene ratio") +
      ylim(c(0,11)) +
      coord_flip() +
      scale_x_discrete(breaks=pathways.dataframe$Pathway.num, 
                 labels=pathways.dataframe$Pathway,
                 position = "top") +
      theme(plot.title = element_text(color="black", size=18, face="bold.italic"),
      plot.subtitle = element_text(color="black", size=16, face="italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = 'black', size=13, hjust =1, family = "mono"), 
      axis.title.x = element_text(face = "bold", color = "black", size = 14),
      axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=13, family = "mono"),
      axis.title.y = element_text(face = "bold", color = "black", size = 14),
      legend.text = element_text(color = "black", size = 12),
      legend.title = element_text(face = "bold", color = "black", size = 14),
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
      labs(title = 'Upregulated Pathways in Reactome and GO Biological Process', 
         x = "Pathways", 
         y = "-Log10(pvalue)") 
    
    assign(paste('PP',cell_type,sep=''),PP)
    
    pdf(paste(outdir,'PP_',cell_type,'.pdf',sep=''),  width=11, height=9)
    print(PP)
    dev.off()    
}


for (cell_type in all.cell.types) {
  HM <- image_read_pdf(paste(outdir,"HM_DS_",cell_type,".pdf", sep=''),density = 160)
  PP <- image_read_pdf(paste(outdir,"PP_",cell_type,".pdf", sep=''), density = 160)
  im = image_append(c(HM, PP))
  image_write(im, path = paste(outdir,"HM_PP_",cell_type,".pdf", sep=''), format = "pdf")
  image_write(im, path = paste(outdir,"HM_PP_",cell_type,".tiff", sep=''), format = "tiff")
}

```


```{r pathway down}
######### Downregulated pathways #######
dat = 'Final_GSE124263_GSE112013_iNOA'
dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'_iNOA/', sep ='')
Seurat.object = iNOA_Final_GSE124263_GSE112013_iNOA.integrated

Seurat.object$cell_type = Seurat.object$celltype
Idents(Seurat.object) <- "celltype.cond"
DefaultAssay(Seurat.object) <- "RNA"

CND = 'down'
cell_types= c("LEY", "MYD", "SRT", "MCR","END", "STRO")
database = c('Reactome_2016','GO_Biological_Process_2018')
gene.l = list()
path.l = list()
gene.list = character()
indir = paste(dir, "enrichR/", sep='')
outdir = paste(dir, "Pathways/", sep='')
for (cell_type in cell_types) {
  print(cell_type)
  gene.list = character()
  enrichR.file = paste(indir,'enrichR_DGE_',cell_type,'_',CND,'.xlsx', sep='')
  
  
  enrichR.table = data.frame()
  for (dat in database) {
    Table <- read.xlsx(xlsxFile = enrichR.file, 
                       sheet = dat, 
                       startRow = 1, 
                       colNames = TRUE,
                       rowNames = TRUE, 
                       detectDates = FALSE, 
                       skipEmptyRows = TRUE,
                       skipEmptyCols = TRUE,
                       na.strings = "NA", 
                       fillMergedCells = FALSE)
    
    enrichR.table = rbind(enrichR.table , Table)
  }
  p = row.names(enrichR.table[enrichR.table$Adjusted.P.value < 0.05,])
  path.l[[cell_type]] = p
  
  for (pathway in p) {
    gene.list.tmp <- unlist(strsplit(enrichR.table[pathway,]$Genes, ';'))
    gene.list = c(gene.list, gene.list.tmp)
    
  }
  gene.l[[cell_type]] = unique(gene.list)
  write.table(gene.list, paste(outdir,'DW_',cell_type,'.txt',sep=''),quote=T, row.names=F, col.names=F)
}


outersect <- function(x, y) {
  sort(c(x[!x%in%y],
         y[!y%in%x]))
}

single.l=list()
only2.l = list()
for (i in cell_types){
    only1 = setdiff(gene.l[[i]],unique(as.character(unlist(gene.l[outersect(cell_types[1:5],i)]))))
    single.l[[i]] = only1
    for (j in cell_types) {
        if (j!=i) {
            only2 = setdiff(unique(as.character(unlist(gene.l[c(i,j)]))),
                            unique(as.character(unlist(gene.l[outersect(cell_types[1:5],c(i,j))]))))
            only2.l[[paste(i,j)]] = only2
        }
        }
}
prova = setdiff(unique(as.character(unlist(gene.l))),
                as.character(unlist(single.l)))
prova2 = setdiff(unique(as.character(unlist(gene.l))),
                 unique(as.character(unlist(only2.l))))

length(prova2)



order.2.plot = c('LEY_iNOA', 'LEY_CTL', 'MYD_iNOA','MYD_CTL','SRT_iNOA','SRT_CTL',
              'END_iNOA','END_CTL','MCR_iNOA','MCR_CTL','TCL_iNOA',
             'STRO_iNOA','STRO_CTL','UND_iNOA','UND_CTL')

#order.2.plot = c('LEY_iNOA', 'LEY_CTL', 'MYD_iNOA','MYD_CTL','SRT_iNOA','SRT_CTL',
#              'END_iNOA','END_CTL','MCR_iNOA','MCR_CTL','TCL_iNOA','TCL_CTL',
#             'STRO_iNOA','STRO_CTL','UND_iNOA','UND_CTL')

#order.2.plot = c('LEY_iNOA', 'LEY_OA', 'MYD_iNOA','MYD_OA','SRT_iNOA','SRT_OA')

levels(Seurat.object) <- order.2.plot
DefaultAssay(Seurat.object) <- "RNA"

#markers.to.plot <- AgeGenesClustersTable_dw$gene.name[AgeGenesClustersTable_dw$cluster==3]
dp_down = DotPlot(Seurat.object, 
        col.min = -1,
        col.max = 1,
        features = prova2, 
        cols = c("lightblue", "red"), 
        dot.scale = 8) +
scale_size_continuous(range = c(0.2,5)) +
RotatedAxis() +
coord_flip() +
theme(axis.text.x = element_text(angle = 45, face = "bold", color = "black", size=15,vjust =1.1), 
      axis.title.x = element_text(face = "bold", color = "black", size = 20),
      axis.text.y = element_text(angle = 0, color = "black", size=11, family = 'mono'), 
      axis.title.y = element_text(face = "bold", color = "black", size = 20),
      legend.text = element_text(face = "bold", color = "black", size = 16),
      legend.position="right",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(x = "Genes", y = "Cell type")

dp_down
pdf(paste(outdir,'DW_genes_exp.pdf',sep=''),width=8, height=19)
plot(dp_down)
dev.off()


p = rev(as.character(sort(Reduce(intersect, path.l))))
pathways.dataframe = data.frame(cellType = character(),
                                Pathway = character(),
                                gene.ratio = numeric(),
                                p.value = numeric(),
                                p.value.adj = numeric(),
                                stringsAsFactors=FALSE)

all.cell.types= cell_types
database = c('Reactome_2016','GO_Biological_Process_2018')
N=length(p)
CND = 'DOWN'
fx <- function(x) eval(parse(text=enrichR.table$Overlap[x]))
fx <- function(x) eval(parse(text=enrichR.table[x,]$Overlap))

for (cell_type in all.cell.types) {
  print(cell_type)
  enrichR.file = enrichR.file = paste(indir,'enrichR_DGE_',cell_type,'_',CND,'.xlsx', sep='')
  enrichR.table = data.frame()
  for (dat in database) {
    Table <- read.xlsx(xlsxFile = enrichR.file, 
                       sheet = dat, 
                       startRow = 1, 
                       colNames = TRUE,
                       rowNames = TRUE, 
                       detectDates = FALSE, 
                       skipEmptyRows = TRUE,
                       skipEmptyCols = TRUE,
                       na.strings = "NA", 
                       fillMergedCells = FALSE)
    
    enrichR.table = rbind(enrichR.table , Table)
  }
  
  pathways.dataframe.entry = data.frame(cellType = rep(cell_type, length(p)),
                                        Pathway = p,
                                        gene.ratio = sapply(p, fx),
                                        p.value = enrichR.table[p,]$P.value,
                                        p.value.adj = enrichR.table[p,]$Adjusted.P.value)
  pathways.dataframe <- rbind(pathways.dataframe, pathways.dataframe.entry)       
}

removeHomo <- function(str, stopwords='Homo') {
  x <- unlist(strsplit(as.character(str), " "))
  paste(x[!x %in% stopwords], collapse = " ")
}
removeSapiens <- function(str, stopwords='sapiens') {
  x <- unlist(strsplit(as.character(str), " "))
  paste(x[!x %in% stopwords], collapse = " ")
}
pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeHomo)
pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeSapiens)


pathways.dataframe$Pathway <- factor(x = pathways.dataframe$Pathway, 
                                     levels = sort(unique(pathways.dataframe$Pathway),decreasing=T))


coloriamo = colors_Alf[all.cell.types,]$cols

patplot = ggplot(data=pathways.dataframe, aes(x=Pathway, y=-log10(p.value), fill=cellType)) +
geom_bar(stat="identity", position=position_dodge()) +
scale_fill_manual(values=as.character(coloriamo)) +
coord_flip() +
scale_x_discrete(position = 'top') + 
theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 90, face = "bold", color = 'black', size=13, hjust =1), 
          axis.title.x = element_text(face = "bold", color = "black", size = 18),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=9, family = 'mono'),
          axis.title.y = element_text(face = "bold", color = "black", size = 18),
          legend.text = element_text(face = "bold", color = "black", size = 14),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid"))

pdf(paste(outdir,'DW_pathPLOT.pdf',sep=''),width=12, height=18)
plot(patplot)
dev.off()

patplot

layout <- '
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
'

p1 = wrap_plots(B = patplot, A = dp_down, design = layout) + plot_layout(guides = 'collect')

pdf(paste(outdir,'Figure7_main.pdf',sep=''),width=20, height=18)
p1
dev.off()

im = image_read_pdf(paste(outdir,"Figure7_main.pdf",sep=''),density = 140)
image_write(im, path = paste(outdir,"Figure7_main.pdf.tiff", sep=''), format = "tiff")

col = as.character(colors_Alf[names(gene.l),]$cols)

venn::venn(gene.l, simplify=TRUE, opacity = 0.3, box = FALSE, ilab=TRUE, zcolor = col, ilcs = 1.5, sncs = 2)

pdf(paste(outdir,'Figure7_venn.pdf',sep=''),width=7, height=6.5)
print(venn::venn(path.l, simplify=TRUE, opacity = 0.3, box = FALSE, elipse = T, ilab=TRUE, zcolor = col, ilcs = 1.5, sncs = 2.5))
dev.off()
im = image_read_pdf(paste(outdir,"Figure7_venn.pdf",sep=''),density = 140)
image_write(im, path = paste(outdir,"Figure7_venn.pdf.tiff", sep=''), format = "tiff")

```


```{r singleR}
#BiocManager::install("SingleR")
BiocManager::install("celldex")
library
matrix = GetAssayData(final_int_obj)
write.table(matrix, file = paste(dat, 'ScaleData_matrix_85.tsv',sep =''), quote = F)

# select Tcells
Tcells = subset(final_int_obj, idents = 'TCL')
table(Tcells$source)
table(Tcells$orig.ident)
Tcells[[]]
matrixTcells = GetAssayData(Tcells)
write.table(matrixTcells, file = paste(dat, 'ScaleData_matrix_Tcells_85.tsv',sep =''), quote = F)
write.table(matrixTcells, file = paste( 'ScaleData_matrix_Tcells_85.tsv',sep =''), quote = F)
dat
Tcells_labels =read.table("./SingleR_testis_TCL.txt")
cells_labels =read.table("./SingleR_testis.txt")
Tcells_labels =read.table('SingleR_testis_TCL_85.txt')
Tcells$singleR = Tcells_labels$labels

length(cells_labels)
final_int_obj$SingleR <- cells_labels$labels

DimPlot(final_int_obj, group.by = 'SingleR') + scale_colour_manual(values = rainbow(25))
Tcells$SingleR <- Tcells_labels$labels
DimPlot(Tcells, group.by = 'singleR', split.by = 'Donor', pt.size = 3)

DimPlot(subset(final_int_obj, idents = 'MCR'), group.by = 'SingleR', split.by = 'condition', pt.size = 3)
#DimPlot(subset(final_int_obj, idents = 'LEY'), group.by = 'SingleR', split.by = 'condition', pt.size = 1)
#DimPlot(subset(final_int_obj, idents = 'MYD'), group.by = 'SingleR', split.by = 'condition', pt.size = 1)

```



*Integration GSE142585 + GSE124263 + GSE112013 + OSR*
```{r PD}
dataset = 'GSE112013'
counts_file_list[['GSE112013']] = '/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/GSE112013_HealthyAdults_Combined_UMI_table.txt'
start_time <- Sys.time()
matrix_sparse <- as.sparse(read.table(counts_file_list[[dataset]], 
                                      sep="\t", 
                                      header = T, 
                                      quote = "", 
                                      row.names = 1))
end_time <- Sys.time()
end_time - start_time
d_Genes = row.names(matrix_sparse)
# GRCh38  Cell Ranger software (v3.0.0)
gtf = rtracklayer::import.gff('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/datasets/Homo_sapiens.GRCh38.84.gtf')
gtf_df = as.data.frame(gtf)[,c("gene_id", "gene_name")]
# gene dataset
gtf_dictionary = gtf_df %>% distinct(gene_id, gene_name, .keep_all = TRUE)
wrong_genes = setdiff(d_Genes, gtf_dictionary$gene_name)
if (length(wrong_genes) != 0) {print("Check dataset genes")} 
d_Genes[d_Genes %in% wrong_genes] = gsub(wrong_genes, pattern = "\\.1", replacement = '')
wrong_genes2 = setdiff(d_Genes, gtf_dictionary$gene_name)
if (length(wrong_genes2) != 0) {print("Check dataset genes again")}

#conversion
start_time <- Sys.time()
conv = sapply(d_Genes, Gene_conversion, gtf_dictionary = gtf_dictionary, gtf_OSR_dictionary = gtf_OSR_dictionary)
end_time <- Sys.time()
end_time - start_time
gene = data.frame(old_name = as.character(names(conv)), new_name = as.character(conv), stringsAsFactors = FALSE)
for (inx in as.numeric(row.names(gene[grep(';',conv),]))) {
  if (length(gtf_OSR_dictionary[gtf_OSR_dictionary$gene_id %in% d_Genes[inx], "gene_name"]) > 0) {
    gene[inx, "new_name"] = gtf_OSR_dictionary[gtf_OSR_dictionary$gene_id %in% d_Genes[inx], "gene_name"]
  } else {
    print(inx)
    gene[inx, "new_name"] = gene[inx, "old_name"]}
}
gene[grep(';',conv),]
p1 = venn::venn(list(GSE112013 = gene$old_name, OSR=Tascini_Genes))
p2 = venn::venn(list(GSE112013 = gene$new_name, OSR=Tascini_Genes))

matrix_sparse_c = matrix_sparse
row.names(matrix_sparse_c) = gene$new_name
testis_GSE112013.raw <- CreateSeuratObject(counts       = matrix_sparse_c, 
                                           project      = dataset, 
                                           min.cells    = 5, 
                                           min.features = 200)
testis_GSE112013.raw
#select somatic cells
load(paste(data.path, 'pz.lit.somatic.best', sep=''))
whichcell = row.names(pz.lit.somatic.best[[]])
testis_GSE112013.raw.somatic = subset(testis_GSE112013.raw, cells = whichcell)

testis_GSE112013.raw.somatic$development = 'adult'
testis_GSE112013.raw.somatic$condition = 'CTL'
testis_GSE112013.raw.somatic$GEO = 'GSE112013'
save(testis_GSE112013.raw.somatic, file = paste(SO_dir, "testis_GSE112013.raw.somatic", sep =''))
D1_cells = row.names(testis_GSE112013.raw.somatic[[]])[grep("Donor1", row.names(testis_GSE112013.raw.somatic[[]]))]
D2_cells = row.names(testis_GSE112013.raw.somatic[[]])[grep("Donor2", row.names(testis_GSE112013.raw.somatic[[]]))]
D3_cells = row.names(testis_GSE112013.raw.somatic[[]])[grep("Donor3", row.names(testis_GSE112013.raw.somatic[[]]))]

testis_GSE112013_D1.somatic = subset(testis_GSE112013.raw.somatic, cells = D1_cells); 
testis_GSE112013_D1.somatic$source = "GSE112013_adult_Donor1"
testis_GSE112013_D2.somatic = subset(testis_GSE112013.raw.somatic, cells = D2_cells); 
testis_GSE112013_D2.somatic$source = "GSE112013_adult_Donor2"
testis_GSE112013_D3.somatic = subset(testis_GSE112013.raw.somatic, cells = D3_cells); 
testis_GSE112013_D3.somatic$source = "GSE112013_adult_Donor3"

row.names(testis_GSE112013_D3.somatic)[grep("^SEL",row.names(testis_GSE112013_D3.somatic))]
```

```{r}
load(paste(SO_dir, "testis_GSE112013.raw.somatic", sep =''))
testis_GSE112013.raw.somatic = get(load(paste(SO_dir, "testis_GSE112013.raw.somatic", sep ='')))

D1_cells = row.names(testis_GSE112013.raw.somatic[[]])[grep("Donor1", row.names(testis_GSE112013.raw.somatic[[]]))]
D2_cells = row.names(testis_GSE112013.raw.somatic[[]])[grep("Donor2", row.names(testis_GSE112013.raw.somatic[[]]))]
D3_cells = row.names(testis_GSE112013.raw.somatic[[]])[grep("Donor3", row.names(testis_GSE112013.raw.somatic[[]]))]

testis_GSE112013_D1.somatic = subset(testis_GSE112013.raw.somatic, cells = D1_cells); 
testis_GSE112013_D1.somatic$source = "GSE112013_adult_Donor1"
testis_GSE112013_D2.somatic = subset(testis_GSE112013.raw.somatic, cells = D2_cells); 
testis_GSE112013_D2.somatic$source = "GSE112013_adult_Donor2"
testis_GSE112013_D3.somatic = subset(testis_GSE112013.raw.somatic, cells = D3_cells); 
testis_GSE112013_D3.somatic$source = "GSE112013_adult_Donor3"


dictionary = list()
dictionary[["Donor1"]][["1"]] = "14515X1"
dictionary[["Donor1"]][["2"]] = "14606X1"
dictionary[["Donor2"]][["1"]] = "14737X1"
dictionary[["Donor2"]][["2"]] = "14737X2"
dictionary[["Donor3"]][["1"]] = "14737X3"
dictionary[["Donor3"]][["2"]] = "14737X4"

# Donor1
replicati = c("1", "2")
Donor = c("Donor1", "Donor2", "Donor3")
for (d in Donor) {
  print(d)
  cells = row.names(testis_GSE112013.raw.somatic[[]])[grep(d, row.names(testis_GSE112013.raw.somatic[[]]))]
  SO_Dsubset = subset(testis_GSE112013.raw.somatic, cells = cells)
  barcodes_d = row.names(SO_Dsubset[[]])
  for (r in replicati) {
    print(r)
    barcodes_d_r =  barcodes_d[grep(paste(".",r,"$", sep=''), barcodes_d)]
    cleanbarcode_d_r = str_sub(barcodes_d_r, start = 8, end = -3)
    write.table(x = cleanbarcode_d_r, 
                file = paste(dictionary[[d]][[r]],'_somatic_whitelist.txt', sep=''), 
                sep ='\t', 
                col.names = FALSE, row.names = FALSE, quote = FALSE)
  }
}

```


```{r integrate iNOA Guo and MW}
############# final integration #########
# load function
source('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/integration_function.R')
options(future.globals.maxSize = 3145728000) 

dat = 'Final_GSE124263_GSE112013_iNOA'

dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'/', sep ='')

load(paste(SO_dir, "iNOA_GSE124263_2D_TCLreassigned.integrated", sep =''))
iNOA_GSE124263_2D_TCLreassigned.integrated = final_int_obj

somatic.list = c(SplitObject(iNOA_GSE124263_2D_TCLreassigned.integrated, split.by = "source"),
                 list(testis_GSE112013_D1.somatic, testis_GSE112013_D2.somatic, testis_GSE112013_D3.somatic))
names(somatic.list) = c('iNOA1', 'iNOA2', 'INOA3', 
                        'GSE124263_adult_Donor1', 'GSE124263_adult_Donor2',
                        'GSE112013_adult_Donor1', 'GSE112013_adult_Donor2','GSE112013_adult_Donor3')

print("input_samples:")
print(somatic.list)

int_obj = integration_workflow(somatic.list = somatic.list, 
                               dataset_iNOA_dir = dir, 
                               dataset = dat)

DimPlot(int_obj, label = T)
levels(int_obj)

# select MCR
DefaultAssay(int_obj) <- "RNA" 
MCR = subset(int_obj, idents = '3')
table(MCR$source)
table(MCR$orig.ident)
head(MCR[[]])
matrixTcells = GetAssayData(MCR)
write.table(matrixTcells, file = paste(dat,'_ScaleData_matrix_MCR_allintegrated.tsv',sep =''), quote = F)
getwd()
MACRO.pred = read.table(file = "/Users/tascini.annasofia/SingleR_testis_FinalIntegration_MCR.tsv", quote = "")

WhichTcells = row.names(MACRO.pred[MACRO.pred$labels %in% c("T_cells", "NK_cell"),])
WhichTcells_f = WhichTcells[! WhichTcells %in% WhichTcells[grep("GSE",WhichTcells)]]
DimPlot(int_obj, cells.highlight = WhichTcells_f)

metadata = int_obj[[]]
#table(metadata$seurat_clusters)
metadata$newIDENT = as.character(metadata$seurat_clusters)
metadata[WhichTcells_f,]$newIDENT = 6
int_obj$Tcell_reassigned = factor(metadata$newIDENT)
table(int_obj$Tcell_reassigned)
Idents(int_obj) <- "Tcell_reassigned"

DimPlot(int_obj, label = T, order = T)
table(Idents(int_obj))

cell_type <- c( "LEY",
                "MYD", 
                "END",
                "MCR",
                "SRT",
                "STRO",
                "TCL")

library(enrichR)
DefaultAssay(int_obj) <- "RNA" 
final_int_obj = integration_postprocessing(SO = int_obj,
                                           new_cluster_id = cell_type, 
                                           dataset = dat, 
                                           dataset_iNOA_dir = dir,
                                           res = 0.2, 
                                           c1 = "iNOA", c2 = "CTL",
                                           colors_Alf = colors_Alf)
save(final_int_obj, file = paste(SO_dir,dat,'.integrated', sep =''))

```

```{r Final UMAP and Ncells}
head(final_int_obj[[]])
final_int_obj$condition = revalue(final_int_obj$condition, c("iNOA"="iGCA", "CTL"="CTL"))
final_int_obj$condition = factor(final_int_obj$condition, levels = c("iGCA", "CTL"))

ND = 8

DimPlot(final_int_obj, group.by = 'condition', pt.size = 0.2) + scale_color_manual(values = c('orange', 'dodgerblue2'))
ggsave(filename = paste(dir, 'UMAP_bycondition.png', sep=''),
       width = 10, height = 9,
       units = 'cm')
DimPlot(final_int_obj, group.by = 'source', order = TRUE) + scale_color_manual(values = rainbow(ND))
ggsave(filename = paste(dir, 'UMAP_byDONOR.png', sep=''),
       width = 15, height = 9,
       units = 'cm')
DimPlot(final_int_obj, group.by = "celltype", split.by = 'condition')
table(Idents(final_int_obj))
final_int_obj[[]]

A <- as.data.frame(prop.table(table(final_int_obj$celltype, 
                                    final_int_obj$condition),margin = 2))
A2 <- as.data.frame(table(final_int_obj$celltype, 
                          final_int_obj$condition),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj$celltype),]$cols)
N <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
  geom_bar(stat="identity",  color="black", position="fill") + 
  scale_fill_manual(values = col) +
  theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=12, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 14),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
        axis.title.y = element_text(face = "bold", color = "black", size = 14),
        legend.title = element_text(size = 0),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = "left",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "", y = "fraction of cells")

A <- as.data.frame(prop.table(table(final_int_obj$celltype, 
                                    final_int_obj$source),margin = 2))
A2 <- as.data.frame(table(final_int_obj$celltype, 
                          final_int_obj$source),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj$celltype),]$cols)
N2 <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
  geom_bar(stat="identity",  color="black", position="fill") + 
  scale_fill_manual(values = col) +
  theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=8, hjust = .5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 14),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
        axis.title.y = element_text(face = "bold", color = "black", size = 14),
        legend.title = element_text(size = 0),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = "left",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "", y = "fraction of cells")
N + N2 + plot_layout(guides = "collect")

pdf(paste(dir,'N_iNOAvsCTL.pdf',sep=''),width=15, height=6)
print(N + N2 + plot_layout(guides = "collect"))
dev.off()



Idents(final_int_obj) <- "celltype"

tail(final_int_obj[[]])
A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$condition),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                                    final_int_obj$condition),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj),]$cols)
N <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
geom_bar(stat="identity",  color="black", position="fill") + 
scale_fill_manual(values = col) +
theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=12, hjust =.5), 
      axis.title.x = element_text(face = "bold", color = "black", size = 14),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
      axis.title.y = element_text(face = "bold", color = "black", size = 14),
      legend.title = element_text(size = 0),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position = "left",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(x = "", y = "fraction of cells")

A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$source),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                                    final_int_obj$source),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj),]$cols)
N2 <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
geom_bar(stat="identity",  color="black", position="fill") + 
scale_fill_manual(values = col) +
#scale_x_discrete(labels=c('GSE124263_adult_Donor1'='GSE124263_#1', 
 #                         'GSE124263_adult_Donor2' = 'GSE124263_#2', 
#                          'iNOA donor 1' = 'iNOA#1', 
 #                         'iNOA donor 2' ='iNOA#2', 
  #                        'iNOA donor 3' ='iNOA#3')) +
theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=8, hjust = .5), 
      axis.title.x = element_text(face = "bold", color = "black", size = 14),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
      axis.title.y = element_text(face = "bold", color = "black", size = 14),
      legend.title = element_text(size = 0),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position = "left",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(x = "", y = "fraction of cells")
N + N2 + plot_layout(guides = "collect")

pdf(paste(dir,'N_iNOAvsCTL.pdf',sep=''),width=15, height=6)
print(N + N2 + plot_layout(guides = "collect"))
dev.off()

table(final_int_obj$source)
```





```{r integration workflow 3datasets + iNOA}
source('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/integration_function.R')
options(future.globals.maxSize = 12145728000) 

dat = 'Final_GSE124263_GSE142585_GSE112013'
dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'/', sep='')

sl = list()
load(paste(SO_dir, "adult.somatic.list", sep =''))
load(paste(SO_dir, "GSE142585.somatic.integrated", sep =''))
testis_GSE142585= SplitObject(GSE142585.somatic.integrated, split.by = "source")$GSE142585
DefaultAssay(testis_GSE142585) <- "RNA"
table(testis_GSE142585$orig.ident)
Donor = str_sub(testis_GSE142585$orig.ident, start = 1, end = 6)
Donor = str_replace(Donor, "Human", "Donor")
table(Donor)
testis_GSE142585$Donor = Donor
Idents(testis_GSE142585) <- "Donor"
GSE142585_adult_Donor1 = subset(testis_GSE142585, idents = "Donor1")
GSE142585_adult_Donor2 = subset(testis_GSE142585, idents = "Donor2")
GSE142585_adult_Donor3 = subset(testis_GSE142585, idents = "Donor3")
GSE142585_adult_Donor4 = subset(testis_GSE142585, idents = "Donor4")

list585 = list(GSE142585_adult_Donor1= GSE142585_adult_Donor1,
               GSE142585_adult_Donor2 = GSE142585_adult_Donor2,
               GSE142585_adult_Donor3= GSE142585_adult_Donor3,
               GSE142585_adult_Donor4 = GSE142585_adult_Donor4)

for (i in names(list585)) {
  list585[[i]]$source = paste(list585[[i]]$source,list585[[i]]$Donor)
}

sl = c(adult.somatic.list, list585)
merge_obj = merge(x = sl[[1]], y = sl[2:length(names(sl))])
merge_obj[["percent.mt"]] <- PercentageFeatureSet(merge_obj, pattern = "^MT-")
merge_obj <- NormalizeData(merge_obj, normalization.method = "LogNormalize", scale.factor = 10000)
merge_obj <- FindVariableFeatures(merge_obj, selection.method = "vst", nfeatures = 2000)
merge_obj <- ScaleData(merge_obj)
merge_obj <- RunPCA(merge_obj, features = VariableFeatures(object = merge_obj))
ElbowPlot(merge_obj)
# cluster the cells 
nPC = 20; res = 0.2
merge_obj <- FindNeighbors(merge_obj, dims = 1:nPC)
merge_obj <- FindClusters(merge_obj, resolution = res)
# UMAP
merge_obj <- RunUMAP(merge_obj, dims = 1:nPC)
DimPlot(merge_obj, group.by = "source")

int_obj = integration_workflow(somatic.list = sl, 
                               dataset_iNOA_dir = dir, 
                               dataset = dat)
library(rcartocolor)
nColor <- length(unique(int_obj$source))
GEO = str_sub(int_obj$source,1,9)
GEO[GEO == "iNOA dono"] = "iGCA"
table(GEO)
int_obj$GEO = GEO
int_obj$GEO = factor(int_obj$GEO, levels = c("iGCA", "GSE142585","GSE112013", "GSE124263"))
col = (carto_pal(nColor, "Vivid")); names(col) <- unique(int_obj$source)
DimPlot(int_obj, label = T)
levels(int_obj)


cell_type <-  c( "MYD",
                      "LEY", 
                      "END",
                      "MCR",
                      "STRO",
                      "SRT",
                      "TCL",
                      "UND")


library(enrichR)
final_int_onj = integration_postprocessing(SO = int_obj,
                                           new_cluster_id = cell_type, 
                                           dataset = dat, 
                                           dataset_iNOA_dir = dir,
                                           res = 0.2, 
                                           colors_Alf = colors_Alf)
                           
DimPlot(final_int_onj, label = T)
levels(final_int_onj )

```


```{r integration workflow 2datasets + iNOA}

source('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/integration_function.R')
options(future.globals.maxSize = 3145728000) 

dat = 'GSE124263_GSE112013'
dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'_iNOA/', sep ='')

sl = list()
sl = list(iNOA_donor1, iNOA_donor2, iNOA_donor3,
          testis_GSE124263.raw.somatic, testis_GSE112013.somatic)
names(somatic.list) <- c('iNOA1', 'iNOA2','iNOA3', "GSE124263", "GSE112013")

int_obj = integration_workflow(somatic.list = sl, 
                               dataset_iNOA_dir = dir, 
                               dataset = dat)

DimPlot(int_obj, label = T)
levels(int_obj)
```

```{r}
dat
cell_type <-  c("LEY", "MYD","END","MCR","STRO", "UND","SRT","TCL","UND")

library(enrichR)
final_int_obj = integration_postprocessing(SO = int_obj,
                                           new_cluster_id = cell_type, 
                                           dataset = dat, 
                                           dataset_iNOA_dir = dir,
                                           res = 0.2, 
                                           colors_Alf = colors_Alf)
                           
DimPlot(final_int_obj, label = T)
levels(final_int_obj)
save(final_int_onj, file = paste(SO_dir,'iNOA_',dat,'.integrated', sep =''))
```

```{r OA testis}
############### OA ###############
testis_somatic_OA = get(load("../testis_somatic_OA"))

testis_OA.somatic = testis_somatic_OA
testis_OA.somatic$condition <- "OA"
testis_OA.somatic$source <- "OA donor"
head(testis_OA.somatic[[]])
testis_OA.somatic <- RenameCells(testis_OA.somatic, add.cell.id = 'OA_')

dat = 'Final_OA'
dir = paste('/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/',dat,'_iNOA/', sep ='')

somatic.list = list()
somatic.list = list(iNOA_donor1, iNOA_donor2, iNOA_donor3,
          testis_OA.somatic)
names(somatic.list) <- c('iNOA1', 'iNOA2','iNOA3', "OA")

dataset_iNOA_dir = dir
dir.create(dataset_iNOA_dir, recursive = TRUE)
res = 0.2
for (i in 1:length(somatic.list)) {
  somatic.list[[i]][["percent.mt"]] <- PercentageFeatureSet(somatic.list[[i]], 
                                                              pattern = "^MT-")
  somatic.list[[i]] <-  subset(somatic.list[[i]],
                                 subset = nFeature_RNA > min_nFeature_RNA & nFeature_RNA < max_nFeature_RNA & percent.mt < max_percent_MT)
  somatic.list[[i]] <- NormalizeData(somatic.list[[i]], 
                                       verbose = FALSE)
  somatic.list[[i]] <- FindVariableFeatures(somatic.list[[i]], 
                                              selection.method = 'vst',
                                              nfeatures = 2000,
                                              verbose = FALSE)
  }
  print("samples after QC filters")
  print(somatic.list)
  
  # integration
  somatic.anchors <- FindIntegrationAnchors(object.list = somatic.list, dims = 1:10, verbose = TRUE, anchor.features = 2000, k.filter = 50)
  somatic.integrated <- IntegrateData(anchorset = somatic.anchors, dims = 1:10, verbose = TRUE)
  
```

```{r OA testis plots} 
  DefaultAssay(somatic.integrated) <- "RNA"
  all.genes <- row.names(somatic.integrated)
  s.genes <- cc.genes$s.genes
  g2m.genes <- cc.genes$g2m.genes
  somatic.integrated <- CellCycleScoring(somatic.integrated, 
                                         s.features = s.genes, 
                                         g2m.features = g2m.genes, 
                                         set.ident = TRUE,
                                         verbose = FALSE)
  
  DefaultAssay(somatic.integrated) <- "integrated"
  
  # SCALE and regress data
  somatic.integrated <- ScaleData(somatic.integrated, 
                                  verbose = FALSE, 
                                  vars.to.regress = c("percent.mt", "nFeature_RNA"), 
                                  features = all.genes)
  # Regress for Cell Cycle
  somatic.integrated <- ScaleData(somatic.integrated, 
                                  verbose = FALSE,
                                  vars.to.regress = c("S.Score", "G2M.Score"), 
                                  features = all.genes) 
  
  somatic.integrated <- RunPCA(somatic.integrated, npcs = 30, verbose = FALSE)
  
  E = ElbowPlot(somatic.integrated)
  ggsave(filename = paste(dataset_iNOA_dir,"ElbowPlot_",dat,".png",sep=''),
         plot = E,
         width = 10, height = 7,
         units = 'cm')
  
  #UMAP
  somatic.integrated <- RunUMAP(somatic.integrated, reduction = "pca", dims = 1:20, verbose = FALSE)
  # clusterinig
  somatic.integrated <- FindNeighbors(somatic.integrated, dims = 1:nPC, verbose = FALSE)
  somatic.integrated <- FindClusters(somatic.integrated, resolution = res, verbose = FALSE)
  somatic.integrated <- RunUMAP(somatic.integrated, dims = 1:nPC, verbose = FALSE)
  
  
  bys = DimPlot(somatic.integrated, reduction = "umap", label = T, pt.size = 1, split.by = 'source', order = TRUE) + 
    theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
          axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
          axis.title.x = element_text(face = "bold", color = "black", size = 24),
          axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
          axis.title.y = element_text(face = "bold", color = "black", size = 24),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position="top",
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
    labs(subtitle = paste('Res = ', res,', nPC = ',nPC, sep = ''), x = "UMAP 1", y = "UMAP 2") + NoLegend()
  
  ggsave(filename = paste(dataset_iNOA_dir,"UMAP_bysources_allcell_integration_",dat,"_iNOA_res",res,".png",sep=''),
         plot = bys,
         width = 45, height = 15, 
         units = 'cm')  

 filename = paste("integration_",dat,"_iNOA_res",res,".xlsx", sep = '')
  existing_file = filename %in% list.files(path = dataset_iNOA_dir, pattern = ".xlsx")
  if (!existing_file) {
    
    cluster.markers = FindAllMarkers(somatic.integrated, verbose=FALSE)
    top10 <- cluster.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
    write.xlsx(cluster.markers,
               file = paste(dataset_iNOA_dir, filename, sep =''), 
               row.names = TRUE,
               asTable = TRUE)
    
    write.csv(cluster.markers, 
              paste(dataset_iNOA_dir, 'seurat_MG_integration_',dataset,'_iNOA_res',res,'.csv', sep =''))
    
    HM =  DoHeatmap(somatic.integrated, features = top10$gene, angle = 90) + NoLegend() + 
      scale_fill_gradientn(colours = coolwarm(200)) 
    ggsave(filename = paste(dataset_iNOA_dir,"HM_integration_",dat,"_iNOA_res",res,".png",sep=''),
           plot = HM,
           width = 14, height = 18, 
           units = 'cm')
  }
  
  leydig_signature = c('CFD','DLK1','LUM','CALB2')
  myoid_signature = c('ACTA2','MYH11','DES', 'MYL9')
  sertoli_signature = c('FATE1','CITED1','SOX9','AMH','CLDN11')
  macrophage_signature = c('CD14','CD74','HLA-DRA','HLA-DRB1')
  endothelial_signature = c('VWF', "EGFL7",'CD34', 'PRSS23', "RBP7")
  t_cell_signature =  c('GZMA','GZMK','CD2','CCL5','NKG7')
  
  DefaultAssay(somatic.integrated) <- 'RNA'
  pL <- Plot_sign(somatic.integrated,
                  signature= c('CFD','DLK1','LUM'), 
                  operator = mean, title = 'LEY')
  pM <- Plot_sign(somatic.integrated,
                  signature= c('ACTA2','MYH11','DES'), 
                  operator = mean, title = 'MYD')
  pS <- Plot_sign(somatic.integrated,
                  signature= c('FATE1','CITED1','SOX9', 'AMH'), 
                  operator = mean, title = 'SRT')
  pMa <- Plot_sign(somatic.integrated,
                   signature= c('CD14','CD74','HLA-DRA'), 
                   operator = mean, title = 'MCR')
  pE <- Plot_sign(somatic.integrated,
                  signature= c('VWF',"EGFL7", 'PRSS23'), 
                  operator = mean, title = 'END')
  pT <- Plot_sign(somatic.integrated,
                  signature= c('GZMA','CD2','CCL5'), 
                  operator = mean, title = 'TCL')
  pSTRO <- Plot_sign(somatic.integrated,
                     signature= c('RGS5','TPM2','IGFBP5'), 
                     operator = mean, title = 'STRO')
  layout <- '
AABBCC##
AABBCC##
AABBCCDD
EEHHGGDD
EEHHGG##
EEHHGG##
'
  png(paste(dataset_iNOA_dir, "MG_integration_",dat,"_iNOA_res",res,".png", sep=''),  width=1400, height=600)
  print(wrap_plots(A = pL, B = pM, C = pS, D = pSTRO, E = pE, H = pMa, G =pT, design = layout))
  dev.off()
  
DefaultAssay(somatic.integrated) <- 'integrated'

table(somatic.integrated$condition)

DimPlot(somatic.integrated)
```
```{r}

```


```{r OA DGE}

cell_type = c('LEY','MYD','MCR', 'END', 'SRT', 'TCL','STRO', 'UND')
  

myDGE_OA = function(Seurat.object, cell_type, outdir = './') {
  azoospermia.response <- FindMarkers(Seurat.object,
                                      ident.1 = paste(cell_type,"_iNOA", sep=''),
                                      ident.2 = paste(cell_type,"_OA", sep=''), 
                                      verbose = FALSE)
  write.xlsx(azoospermia.response,
             file= paste(outdir,'DGE_', cell_type,'.iNOA.vs.OA.xlsx',sep=''), 
             row.names = T,
             asTable = T)
    return(azoospermia.response)
}
  

  SO = somatic.integrated
  new_cluster_id = cell_type 
  dataset = 'OA'
  dataset_iNOA_dir = dir
  names(new_cluster_id) <- levels(SO)
  SO <- RenameIdents(SO, new_cluster_id)
  
  col = as.character(colors_Alf[levels(SO),]$cols)
  
  png(paste(dataset_iNOA_dir,"UMAP_cellAssigned_",dataset,"_iNOA_res",res,".png",sep=''),  
      width = 500, height = 400)
  print(DimPlot(SO, reduction = "umap", label = TRUE, pt.size = 1, order = TRUE) +
    scale_color_manual(values = (col)) +
    theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
          axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
          axis.title.x = element_text(face = "bold", color = "black", size = 24),
          axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
          axis.title.y = element_text(face = "bold", color = "black", size = 24),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position="right",
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
    labs(x = "UMAP 1", y = "UMAP 2"))
  dev.off()
  
  filename = paste("integration_",dataset,"_iNOA_res",res,"_celltype.xlsx", sep = '')
  existing_file = filename %in% list.files(path = dataset_iNOA_dir, pattern = ".xlsx")
  
  
  if (!existing_file) {
    cluster.markers = FindAllMarkers(SO)
    top10 <- cluster.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
    write.xlsx(cluster.markers,
               file= paste(dataset_iNOA_dir, filename, sep =''), 
               row.names = T,
               asTable = T)
    HM =  DoHeatmap(SO, features = top10$gene, angle = 90) + NoLegend() + 
      scale_fill_gradientn(colours = coolwarm(200)) 
    
    png(paste(dataset_iNOA_dir,"HM_celltype_integration_",dataset,"_iNOA_res",res,".png",sep=''),  width = 1400, height = 1600)
    plot(HM)
    dev.off()
  }
  
  ######## DGE ###########
  dir_DGE = paste(dataset_iNOA_dir,'DGE/', sep='')
  dir.create(dir_DGE)
  
  DefaultAssay(SO) <- "RNA"
  # relabel cells to perform DGE
  SO$celltype.cond <- paste(Idents(SO), SO$condition, sep = "_")
  SO$celltype <- Idents(SO)
  Idents(SO) <- "celltype.cond"
  new_cluster_id_filtered = c('MYD', 'LEY', 'SRT')
  DGEresults = lapply(new_cluster_id_filtered, myDGE_OA, Seurat.object = SO, outdir = dir_DGE)
  names(DGEresults) <- c('MYD', 'LEY', 'SRT')
  filename = paste(dataset_iNOA_dir,"DGEall_integration_",dataset,"_res",res,".xlsx", sep = '')
  write.xlsx(x = DGEresults, file = filename, asTable = T, row.names = T)
  
  ######## enrichment ###########
  dir_enrichR = paste(dataset_iNOA_dir,'enrichR/', sep='')
  dir.create(dir_enrichR)
  
  databases <- listEnrichrDbs()
  # databases to make the enrichment of
  enrich.databases <- c("GO_Biological_Process_2018",
                        "GO_Cellular_Component_2018",
                        "GO_Molecular_Function_2018",
                        "Reactome_2016",
                        "KEGG_2016",
                        "WikiPathways_2016",
                        "BioCarta_2016",
                        "BioPlanet_2019")
  
  
  cell_types = new_cluster_id
  DGE_gene_up_list <- list()
  DGE_gene_dw_list <- list()
  
  for (celltype in new_cluster_id_filtered) {
    DGE_file = paste(dir_DGE,"DGE_", celltype, ".iNOA.vs.OA.xlsx", sep= '')
    A <- read.xlsx(DGE_file)
    
    up.genes <- A$row.names[A$avg_logFC > 0 & A$p_val_adj <= 0.05]
    down.genes <- A$row.names[A$avg_logFC < 0 & A$p_val_adj <= 0.05]
    both.genes <- A$row.names[A$p_val_adj <= 0.05]
    
    DGE_gene_up_list[[celltype]] <- up.genes
    DGE_gene_dw_list[[celltype]] <- down.genes
    
    enrichr.list <- list()
    enrichr.list <- lapply(list(up.genes,down.genes,both.genes), function(x) {
      enrichR::enrichr(genes = x, databases = enrich.databases)
    })
    
    names(enrichr.list) <-  c("up","down","both")
    
    for (i in names(enrichr.list)){
      for (dat in names(enrichr.list[[i]])) {
        enrichr.list[[i]][[dat]] = enrichr.list[[i]][[dat]][,
                                                            c('Term', 
                                                              'Overlap', 
                                                              'P.value', 
                                                              'Adjusted.P.value', 
                                                              'Odds.Ratio', 
                                                              'Combined.Score', 
                                                              'Genes')]
      }
    }
    for (i in 1:length(enrichr.list)) {
      filename = paste(dir_enrichR,'enrichR_DGE_',celltype,'_',names(enrichr.list)[i],'.xlsx', sep='')
      if (!is.null(enrichr.list[[i]])) {write.xlsx(x = enrichr.list[[i]], file = filename, asTable = T)}
    }
  }
  
  
  Idents(SO) <- "celltype"
  SO$condition <- factor(x = SO$condition, levels = c("iNOA", "OA"))
  #SO  = somatic.integrated.new2
  for (g in  c('CD3E','TRAC', 'GZMK')) {
    v = VlnPlot(subset(SO, idents = c('MCR','TCL')), features =g, 
                split.by = 'condition', assay = 'RNA', split.plot = F, pt.size = 0.5) + 
      theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
            plot.subtitle = element_text(color="black", size=16, face="italic"),
            axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
            axis.title.x = element_text(face = "bold", color = "black", size = 18),
            axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
            axis.title.y = element_text(face = "bold", color = "black", size = 18),
            legend.text = element_text(face = "bold", color = "black", size = 12),
            legend.position = c(0.1, 0.92),
            panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
      #geom_boxplot(width = 0.1, outlier.size=1) +
      scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iNOA','CTL'))
    assign(paste('v_', g, sep =''),v)
  }
  v_CD3E | v_TRAC | v_GZMK
  
  png(paste(dataset_iNOA_dir,"MG_Tcell_",dataset,"_iNOA_res",res,".png",sep=''),  width = 1400, height = 600)
  plot(v_CD3E | v_TRAC | v_GZMK)
  dev.off()
  
  
  for (g in  c('HLA-DRA','CD74', 'HLA-DPA1')) {
    v = VlnPlot(subset(SO , idents = c('MCR','TCL')), features =g, 
                split.by = 'condition', assay = 'RNA', pt.size = .5) + 
      theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
            plot.subtitle = element_text(color="black", size=16, face="italic"),
            axis.text.x = element_text(angle = 45, face = "bold", color = "black", size = 16, hjust =1), 
            axis.title.x = element_text(face = "bold", color = "black", size = 18),
            axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
            axis.title.y = element_text(face = "bold", color = "black", size = 18),
            legend.text = element_text(face = "bold", color = "black", size = 12),
            legend.position = c(0.5, 0.92),
            panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
      #geom_boxplot(width = 0.1, outlier.size=1) +
      scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iNOA','CTL'))
    assign(paste('v_', g, sep =''),v)
  }
  
  `v_HLA-DRA` | v_CD74 | `v_HLA-DPA1`
  png(paste(dataset_iNOA_dir,"MG_MCR_",dataset,"_iNOA_res",res,".png",sep=''),  width = 1400, height = 600)
  plot(`v_HLA-DRA` | v_CD74 | `v_HLA-DPA1`)
  dev.off()

  
library(enrichR)
final_int_obj = SO
save(final_int_obj, file = paste(SO_dir,'iNOA_',dat,'.integrated', sep =''))

DimPlot(final_int_obj)

table(final_int_obj$condition)
```

```{r OA plots}
####### N cells ##########

ND = 4

a = DimPlot(final_int_obj, group.by = 'celltype', pt.size = 1, label =T) + scale_color_manual(values = col) + 
    theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
          axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=12, hjust =1), 
          axis.title.x = element_text(face = "bold", color = "black", size = 14),
          axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
          axis.title.y = element_text(face = "bold", color = "black", size = 14),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position="right",
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
    labs(x = "UMAP 1", y = "UMAP 2") + NoLegend()
b = DimPlot(final_int_obj, group.by = 'condition', pt.size = 1) + 
  scale_color_manual(values = c('orange', '#33FF99')) + 
    theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
          axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=12, hjust =1), 
          axis.title.x = element_text(face = "bold", color = "black", size = 14),
          axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
          axis.title.y = element_text(face = "bold", color = "black", size = 14),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position="top",
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
    labs(x = "UMAP 1", y = "UMAP 2")

ggsave(filename = paste(dir, 'UMAP_bycondition.png', sep=''),
       a+b,
       width = 15, height = 8,
       units = 'cm')

DimPlot(final_int_obj, group.by = 'source', order = TRUE) + scale_color_manual(values = rainbow(ND))
ggsave(filename = paste(dir, 'UMAP_byDONOR.png', sep=''),
       width = 15, height = 9,
       units = 'cm')
DimPlot(final_int_obj, group.by = "celltype", split.by = 'condition')
table(Idents(final_int_obj))


A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$condition),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                          final_int_obj$condition),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj),]$cols)
N <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
  geom_bar(stat="identity",  color="black", position="fill") + 
  scale_fill_manual(values = col) +
  scale_x_discrete(labels=c('iGCA', 'OA')) + 
  theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=12, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 14),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
        axis.title.y = element_text(face = "bold", color = "black", size = 14),
        legend.title = element_text(size = 0),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = "left",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "", y = "fraction of cells")

A <- as.data.frame(prop.table(table(Idents(final_int_obj), 
                                    final_int_obj$source),margin = 2))
A2 <- as.data.frame(table(Idents(final_int_obj), 
                          final_int_obj$source),margin = 2)
A$N <- A2$Freq

col = as.character(colors_Alf[levels(final_int_obj),]$cols)
N2 <- ggplot(data=A, aes(x=Var2, y=N, fill=Var1)) +
  geom_bar(stat="identity",  color="black", position="fill") + 
  scale_fill_manual(values = col) +
  #scale_x_discrete(labels=c('GSE124263_adult_Donor1'='GSE124263_#1', 
  #                         'GSE124263_adult_Donor2' = 'GSE124263_#2', 
  #                          'iNOA donor 1' = 'iNOA#1', 
  #                         'iNOA donor 2' ='iNOA#2', 
  #                        'iNOA donor 3' ='iNOA#3')) +
  theme(plot.title = element_text(color="black", size=16, face="bold.italic"),
        axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=8, hjust = .5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 14),
        axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=12),
        axis.title.y = element_text(face = "bold", color = "black", size = 14),
        legend.title = element_text(size = 0),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = "left",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "", y = "fraction of cells")
N + N2 + plot_layout(guides = "collect")

pdf(paste(dir,'N_iNOAvsCTL.pdf',sep=''),width=15, height=6)
print(N + N2 + plot_layout(guides = "collect"))
dev.off()



#COL = row.names(Seurat.object)[grep("^COL",row.names(Seurat.object))]
gene = c("DLK1", "NOTCH2", "IGF1", "IGF2", "INSL3", "HSD17B3",
         "CDKN2A",
         "COL1A1", "COL1A2", "COL4A1", "COL4A2", 
         'H19', 'MEG3', 'MEG8', 'HYMAI', 'PEG3', 'ERAP2', 'RTL1', "KCNQ1OT1")

Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <-"RNA"
Idents(Seurat.object) <- "celltype.cond"

azoospermia.modulated.gene = data.frame()
for (ct in c('LEY','MYD','SRT')) {
  #print(ct)
  azoospermia.response.gene <- data.frame()
  azoospermia.response.gene <- FindMarkers(Seurat.object,
                                           ident.1 = paste(ct,"_iNOA", sep=''),
                                           ident.2 = paste(ct,"_OA", sep=''), 
                                           verbose = FALSE,
                                           feature=gene,
                                           logfc.threshold=0,
                                           min.pct = 0)
  
  azoospermia.response.gene          <- azoospermia.response.gene[gene,] 
  azoospermia.response.gene$geneID   <- rownames(azoospermia.response.gene)
  azoospermia.response.gene$cellType <- ct 
  azoospermia.modulated.gene         <- rbind(azoospermia.modulated.gene, azoospermia.response.gene)  
}

#azoospermia.modulated.gene 
myorder=c('cellType','geneID','avg_logFC','p_val_adj','p_val','pct.1','pct.2')
azoospermia.modulated.gene <- azoospermia.modulated.gene[,myorder]
write.xlsx(azoospermia.modulated.gene,
           file= paste(dir,'azoospermia.modulated.gene.azoospermia.xlsx',sep=''), 
           row.names = FALSE,
           asTable = TRUE)

####### CDKN2A ########
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
Seurat.object_s <- Seurat.object # subset(Seurat.object,  idents = c("LEY", "MYD","SRT"))
#Seurat.object_s = Seurat.object
vp_D <- VlnPlot(Seurat.object_s,
                feature = "CDKN2A",
                #slot = "counts", 
                #log = TRUE,
                split.plot = T,
                pt.size = 2,
                split.by = 'condition') +
  ylab("Expression Level") +
  xlab("") +
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.9),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA')) +
  scale_color_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA'))

fp_D <- FeaturePlot(Seurat.object, 
                    reduction = "umap", 
                    features = c("CDKN2A"), 
                    label = T, 
                    order= T,
                    pt.size = 2,
                    max.cutoff = 'q99',
                    cols = c("lightgrey", "red"),
                    split.by='condition') + theme(legend.position = 'right')

vp_D
fp_D
ggsave(filename = paste(dir,"Featureplot_CDKN2A.png",sep=''),
       plot = fp_D,
       width = 20, height = 10,
       units = 'cm')
ggsave(filename = paste(dir,"Violinplot_CDKN2A.png",sep=''),
       plot = vp_D,
       width = 32, height = 16,
       units = 'cm')

##### DLK1 NOTCH2 #####

#DLK1
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
#Seurat.object_s <- subset(Seurat.object,  idents = c("LEY", "LEY2", "MYD","SRT"))
Seurat.object_s <- subset(Seurat.object,  idents = c("LEY", "MYD"))
#Seurat.object_s = Seurat.object
vp_D <- VlnPlot(Seurat.object_s,
                feature = 'DLK1',
                #slot = "counts", 
                #log = TRUE,
                split.plot = T,
                pt.size = 0,
                split.by = 'condition') +
  ylab("Expression Level") +
  xlab("") +
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.9),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA'))

fp_D <- FeaturePlot(Seurat.object, 
                    reduction = "umap", 
                    features = c('DLK1'), 
                    label = T, 
                    order= T,
                    max.cutoff = 'q99',
                    cols = c("lightgrey", "red"),
                    split.by='condition') + theme(legend.position = 'right')


vp_N <- VlnPlot(Seurat.object_s,
                feature = 'NOTCH2',
                #slot = "counts", 
                #log = TRUE,
                split.plot = T,
                pt.size = 0,
                split.by = 'condition') +
  ylab("Expression Level") +
  xlab("") +
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.9),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA'))

Seurat.object$condition = plyr::revalue(Seurat.object$condition, c("iNOA"="iGCA", "CTL"="OA"))

fp_N <- FeaturePlot(Seurat.object, 
                    reduction = "umap", 
                    pt.size = 1,
                    features = c('NOTCH2'), 
                    label = T, 
                    order= T,
                    max.cutoff = 'q99',
                    cols = c("lightgrey", "red"),
                    split.by='condition') + theme(legend.position = 'right')


fp_blend <- FeaturePlot(Seurat.object, 
                        reduction = "umap", 
                        pt.size = 1,
                        features = c('DLK1','NOTCH2'),
                        label = T, 
                        order= T,
                        blend = T,       
                        max.cutoff = 'q99',
                        cols = c("lightgrey", "red","green"),
                        split.by='condition') + theme(legend.position = 'right')


layout <- '
AACCCC
BBCCCC
' 


pdf(paste(dir,'Figure3_DLK1_NOTCH2_blend.pdf',sep=''),width=19, height=7)
wrap_plots(A = vp_D, B = vp_N, C = fp_blend, design = layout)
dev.off()

im = image_read_pdf(paste(dir,"Figure3_DLK1_NOTCH2_blend.pdf",sep=''),density = 140)
image_write(im, path = paste(dir,"Figure3_DLK1_NOTCH2_blend.tiff", sep=''), format = "tiff")

######## IGF1 IGF2##########
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
#Seurat.object_s <- subset(Seurat.object, idents = c("LEY","LEY2", "MYD","SRT","END","STRO"))
#Seurat.object_s <- subset(Seurat.object, idents = c("LEY", "MYD"))
Seurat.object_s <- subset(Seurat.object, idents = c("LEY", "MYD"))
#

vp_IG1 <- VlnPlot(Seurat.object_s,
                  feature = c('IGF1'),
                  #slot = "counts", 
                  #log = TRUE,
                  split.plot = T,
                  pt.size = 0,
                  split.by = 'condition') +
  ylab("Expression Level") +
  xlab("") +
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.9),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA'))

vp_IG2 <- VlnPlot(Seurat.object_s,
                  feature = c('IGF2'),
                  #slot = "counts", 
                  #log = TRUE,
                  split.plot = T,
                  pt.size = 0,
                  split.by = 'condition') +
  ylab("Expression Level") +
  xlab("") +
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.9),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA'))

Seurat.object$condition = plyr::revalue(Seurat.object$condition, c("iNOA"="iGCA", "CTL"="OA"))
fp_bl <- FeaturePlot(Seurat.object, 
                     reduction = "umap", 
                     pt.size = 1,
                     features = c('IGF1','IGF2'), 
                     label = T, 
                     order= T,
                     blend = T,       
                     max.cutoff = 'q99',
                     #cols = c("lightgrey", "red"),
                     split.by='condition') + theme(legend.position = 'right')

fp_bl
vp_IG1

layout <- '
AA
BB
' 
pdf(paste(dir,'Figure3_IGFs_blend.pdf',sep=''),width=6, height=7)
wrap_plots(A = vp_IG1, B = vp_IG2, design = layout)
dev.off()

im = image_read_pdf(paste(dir,"Figure3_IGFs_blend.pdf",sep=''),density = 140)
image_write(im, path = paste(dir,"Figure3_IGFs_blend.tiff", sep=''), format = "tiff")

#### COL1A1 COL1A2 ######
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
#Seurat.object_s <- subset(Seurat.object,  idents = c("LEY", "LEY2","MYD","STRO"))
Seurat.object_s <- subset(Seurat.object,  idents = c("LEY","MYD"))
#Seurat.object_s <- subset(Seurat.object,  idents = c("LEY", "MYD"))

vp_IG1 <- VlnPlot(Seurat.object_s,
                  feature = c('COL1A1'),
                  #slot = "counts", 
                  #log = TRUE,
                  split.plot = T,
                  pt.size = 0,
                  split.by = 'condition') +
  ylab("Expression Level") +
  xlab("") +
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.9),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA'))

vp_IG2 <- VlnPlot(Seurat.object_s,
                  feature = c('COL1A2'),
                  #slot = "counts", 
                  #log = TRUE,
                  split.plot = T,
                  pt.size = 0,
                  split.by = 'condition') +
  ylab("Expression Level") +
  xlab("") +
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.9),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c('iGCA','OA'))

Seurat.object$condition = plyr::revalue(Seurat.object$condition, c("iNOA"="iGCA", "CTL"="OA"))
fp_bl <- FeaturePlot(Seurat.object, 
                     reduction = "umap", 
                     pt.size = 1,
                     features = c('COL1A1','COL1A2'), 
                     label = T, 
                     order= T,
                     blend = T,       
                     max.cutoff = 'q99',
                     #cols = c("lightgrey", "red"),
                     split.by='condition') + theme(legend.position = 'right')


layout <- '
AACCCC
BBCCCC
' 
pdf(paste(dir,'Figure5_COL1s_blend.pdf',sep=''),width=19, height=7)
wrap_plots(A = vp_IG1, B = vp_IG2, C = fp_bl, design = layout)
dev.off()

im = image_read_pdf(paste(dir,"Figure5_COL1s_blend.pdf",sep=''),density = 140)
image_write(im, path = paste(dir,"Figure5_COL1s_blend.tiff", sep=''), format = "tiff")


######### imprinted genes ##########
relevant_imprinted = c('H19', 'MEG3', 'MEG8','IGF2', 'DLK1', 'HYMAI', 'PEG3', 'ERAP2', 'RTL1', "KCNQ1OT1")
dir.create(paste(dir,'ImprintedGenes/', sep = ''))
dir.create(paste(dir,'ImprintedGenes/VlnPlot_r', sep = ''))
Seurat.object = final_int_obj
DefaultAssay(Seurat.object) <- "RNA"
#so = subset(Seurat.object, idents = c('LEY', 'LEY2','MYD'))
so = subset(Seurat.object, idents = c('LEY','MYD'))

for (gene in relevant_imprinted) {
  so[[gene]] <- FetchData(object = so, vars = gene)
  df= so[[]]
  df$condition = plyr::revalue(df$condition, c("iNOA"="iGCA", "CTL"="OA"))
  v <- ggviolin(df, x = "condition", y = gene, 
                fill = "condition", color = "condition", width = 1,
                add = c('jitter'), add.params = list(size =.5, shape = 1),
                #              add.params = list(size = .5, shape = 20),
                alpha = 0.6,
                facet.by = 'celltype', short.panel.labs = TRUE,
                palette = c('orange', 'dodgerblue2'), 
                error.plot = "crossbar", draw_quantiles = c(0.5), trim = T,
                title = gene, 
                panel.labs.background = list(color = "white", fill = "white", size = 0.5),
                panel.labs.font = list(color = "black", face = "bold", size = 16))
  df$condition
  v = ggpar(v, ylab = 'Expression level', xlab = '')
  v = v + stat_compare_means(aes(group = condition),label = "p.signif", size =  8, 
                             label.x.npc = "center", label.y.npc = 0.85) +
    
    theme(plot.title = element_text(color="black", size=22, face="bold.italic", hjust = 0.5),
          plot.title.position =  'panel',
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 14, hjust =.5), 
          axis.title.x = element_text(face = "bold", color = "black", size = 14),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=14),
          axis.title.y = element_text(face = "bold", color = "black", size = 16),
          legend.text = element_text(face = "bold", color = "black", size = 12),
          legend.position = 'right',
          panel.spacing = unit('-0.1', "lines"),
          #legend.text = c('',''),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid"))
  assign(paste('v','c',gene,sep = '_'),v)
  pdf(paste(dir, 'ImprintedGenes/VlnPlot_r/','vlnPlot_',gene,'.pdf', sep=''), width = 5, height = 4.5)
  print(v)
  dev.off()
  
}

pdf(paste(dir, 'ImprintedGenes/VlnPlot_r/','vlnPlot_maternal.pdf', sep=''), width = 12, height = 4)
print(v_c_H19 + v_c_MEG3 + v_c_MEG8 + plot_layout(guides = "collect", ncol = 3) & 
        theme(legend.position = "right", legend.direction = 'vertical'))
dev.off()

pdf(paste(dir, 'ImprintedGenes/VlnPlot_r/','vlnPlot_paternal.pdf', sep=''), width = 28, height = 4)
print(v_c_IGF2 + v_c_DLK1 + v_c_HYMAI + v_c_PEG3 + v_c_ERAP2 + v_c_RTL1 + v_c_KCNQ1OT1 + plot_layout(guides = "collect", ncol = 7) & 
        theme(legend.position = "right", legend.direction = 'vertical'))
dev.off()


################ INSL3 #################
Seraut.object= final_int_obj

CELL_INSL3 <- WhichCells(Seraut.object, slot = 'counts', expression = INSL3 > 1)
somatic.integrated.new.INSL3 <- subset(Seraut.object, cells = CELL_INSL3)


legendiNOA='iGCA'
legendCTRL='OA'

Seraut.object = somatic.integrated.new.INSL3
Idents(Seraut.object) <- "celltype"

cell_type.2.use = 'LEY'


cell_type = c(paste(cell_type.2.use,'_iNOA',sep=''),
              paste(cell_type.2.use,'_OA',sep=''))
features = c('INSL3')


Cell.subset <- subset(Seraut.object, idents =  cell_type.2.use, invert=F)
#subset.counts <- Cell.subset@assays$RNA@counts
subset.counts <- Cell.subset@assays$RNA@data
gene.counts.dataframe = data.frame(gene = character(),
                                   cell_barcode = character(),
                                   cell_type = character(),
                                   counts = integer(),
                                   stringsAsFactors=FALSE)
for (gene in features) {
  gene.counts.tmp = subset.counts[gene,]
  gene.counts.dataframe.entry = data.frame(gene = rep(gene, length(gene.counts.tmp)),
                                           cell_barcode =  names(gene.counts.tmp),
                                           condition = Seraut.object[[]][names(gene.counts.tmp),
                                                                         'celltype.cond'],
                                           counts = (gene.counts.tmp))
  gene.counts.dataframe <- rbind(gene.counts.dataframe, gene.counts.dataframe.entry)
}
gene.counts.dataframe$condition <- factor(x = gene.counts.dataframe$condition, 
                                          levels = c(paste(cell_type.2.use,"_iNOA",sep=''),paste(cell_type.2.use,"_CTL",sep='')))
#paste(cell_type.2.use,"_CTL",sep='')))

CVP = ggplot(gene.counts.dataframe, aes(x=gene, y = counts, fill=condition)) +
  geom_split_violin() + 
  #ylim(0,6) +
  #scale_y_log10() +
  scale_x_discrete('LEY') +
  #scale_y_continuous(trans='log') +
  scale_fill_manual(values = c('purple', 'orange')) +
  #geom_jitter(shape=16, size = 0.3, position=position_jitter(0.4)) +  
  theme(plot.title = element_text(color="black", size=22, face="bold.italic", hjust = 0.5),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_blank(),
        #          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5),
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.87),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #+
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c(legendiNOA,legendCTRL)) +
  labs(title = cell_type.2.use , 
       x = "Genes", 
       y = "Expression Level") 
CVP = CVP +labs(title= 'INSL3',
                x = "LEY", 
                y = "Expression Level") 
assign('INSL3_INSL3cell',CVP)


Seraut.object= final_int_obj
Idents(Seraut.object) <- "celltype"


features = c('INSL3')


Cell.subset <- subset(Seraut.object, idents =  cell_type.2.use, invert=F)
#subset.counts <- Cell.subset@assays$RNA@counts
subset.counts <- Cell.subset@assays$RNA@data
gene.counts.dataframe = data.frame(gene = character(),
                                   cell_barcode = character(),
                                   cell_type = character(),
                                   counts = integer(),
                                   stringsAsFactors=FALSE)
for (gene in features) {
  gene.counts.tmp = subset.counts[gene,]
  gene.counts.dataframe.entry = data.frame(gene = rep(gene, length(gene.counts.tmp)),
                                           cell_barcode =  names(gene.counts.tmp),
                                           condition = Seraut.object[[]][names(gene.counts.tmp),
                                                                         'celltype.cond'],
                                           counts = (gene.counts.tmp))
  gene.counts.dataframe <- rbind(gene.counts.dataframe, gene.counts.dataframe.entry)
}
gene.counts.dataframe$condition <- factor(x = gene.counts.dataframe$condition, 
                                          levels = c(paste(cell_type.2.use,"_iNOA",sep=''),paste(cell_type.2.use,"_CTL",sep='')))
#paste(cell_type.2.use,"_CTL",sep='')))

CVP = ggplot(gene.counts.dataframe, aes(x=gene, y = counts, fill=condition)) +
  geom_split_violin() + 
  #ylim(0,6) +
  #scale_y_log10() +
  scale_x_discrete('LEY') +
  #scale_y_continuous(trans='log') +
  scale_fill_manual(values = c('purple', 'orange')) +
  #geom_jitter(shape=16, size = 0.3, position=position_jitter(0.4)) +  
  theme(plot.title = element_text(color="black", size=22, face="bold.italic", hjust = 0.5),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_blank(),
        #          axis.text.x = element_text(angle = 0, face = "bold", color = "black", size = 16, hjust =.5),
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 12),
        legend.position = c(0.85, 0.87),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  #+
  #geom_boxplot(width = 0.1, outlier.size=1) +
  scale_fill_manual(values = c('orange', 'dodgerblue2'), labels = c(legendiNOA,legendCTRL)) +
  labs(title = cell_type.2.use , 
       x = "Genes", 
       y = "Expression Level") 
CVP = CVP +labs(title= 'INSL3',
                x = "LEY", 
                y = "Expression Level") 
CVP

assign('INSL3_allcell',CVP) 

INSL3_allcell + plot_spacer() + INSL3_INSL3cell
p1 =INSL3_allcell; p2 =INSL3_INSL3cell;
(p1 + theme(plot.margin = unit(c(0,30,0,0), "pt"))) +
  (p2 + theme(plot.margin = unit(c(0,0,0,30), "pt")))
pdf(paste(dir,'Figure_INSL3.pdf',sep=''),width=8.5, height=3.5)
plot((p1 + theme(plot.margin = unit(c(0,40,0,15), "pt"))) +
       (p2 + theme(plot.margin = unit(c(0,15,0,40), "pt"))))
dev.off()

im = image_read_pdf(paste(dir,'Figure_INSL3.pdf',sep=''),density = 140)
image_write(im, path = paste(dir,'Figure_INSL3.tiff', sep=''), format = "tiff")


#### Pathway UP heatmaps ####
Seurat.object = final_int_obj
Seurat.object$cell_type = Seurat.object$celltype
Idents(Seurat.object) <- "celltype.cond"
DefaultAssay(Seurat.object) <- "RNA"

dir

CND = 'up'
database = c('Reactome_2016','GO_Biological_Process_2018')
indir =  paste(dir,'enrichR/', sep = '')
indir
cell.type_2_plot = c("LEY", "MYD")

pathway_to_plot <- list()
pathway_to_plot[["LEY"]] <- c("Cholesterol biosynthesis Homo sapiens R-HSA-191273",
                              "Extracellular matrix organization Homo sapiens R-HSA-1474244",
                              "Collagen formation Homo sapiens R-HSA-1474290",
                              "Collagen biosynthesis and modifying enzymes Homo sapiens R-HSA-1650814",
                              "collagen fibril organization (GO:0030199)",
                              "IRE1alpha activates chaperones Homo sapiens R-HSA-381070",
                              "regulation of cell migration (GO:0030334)",
                              "regulation of keratinocyte apoptotic process (GO:1902172)")


pathway_to_plot[["MYD"]] <- c("Cholesterol biosynthesis Homo sapiens R-HSA-191273",
                              "type I interferon signaling pathway (GO:0060337)",
                              "regulation of interferon-gamma-mediated signaling pathway (GO:0060334)",
                              "Extracellular matrix organization Homo sapiens R-HSA-1474244",
                              "Collagen biosynthesis and modifying enzymes Homo sapiens R-HSA-1650814",
                              "Scavenging by Class A Receptors Homo sapiens R-HSA-3000480",
                              "regulation of cell differentiation (GO:0045595)",
                              "glutathione derivative metabolic process (GO:1901685)",
                              "negative regulation of cellular macromolecule biosynthetic process (GO:2000113)",
                              "regulation of cell migration (GO:0030334)")


pathway_to_plot[["MCR"]] <- c("MHC class II antigen presentation Homo sapiens R-HSA-2132295",
                              "PD-1 signaling Homo sapiens R-HSA-389948",
                              "Interferon gamma signaling Homo sapiens R-HSA-877300",
                              "regulated exocytosis (GO:0045055)",
                              "regulation of amyloid-beta formation (GO:1902003)",
                              "neutrophil mediated immunity (GO:0002446)",
                              "homotypic cell-cell adhesion (GO:0034109)",
                              "cytokine-mediated signaling pathway (GO:0019221",
                              "production of miRNAs involved in gene silencing by miRNA (GO:0035196)",
                              "regulation of transcription from RNA polymerase II promoter (GO:0006357)",
                              "regulation of protein phosphorylation (GO:0001932)")

pathway_to_plot[["LEY"]] <- c("Extracellular matrix organization Homo sapiens R-HSA-1474244",
  "nuclear-transcribed mRNA catabolic process, deadenylation-independent decay (GO:0031086)",
                              "Collagen biosynthesis and modifying enzymes Homo sapiens R-HSA-1650814",
                              "Scavenging by Class A Receptors Homo sapiens R-HSA-3000480",
                              "cellular response to corticosteroid stimulus (GO:0071384)")


pathway_to_plot[["MYD"]] <- c("Extracellular matrix organization Homo sapiens R-HSA-1474244",
                               "Antigen Presentation: Folding, assembly and peptide loading of class I MHC Homo sapiens R-HSA-983170",
                              "Pre-NOTCH Transcription and Translation Homo sapiens R-HSA-1912408",
                              "Scavenging by Class A Receptors Homo sapiens R-HSA-3000480")

indir = paste(dir, "enrichR/", sep='')
outdir = paste(dir, 'Pathways/', sep='') 
dir.create(outdir)

for (i in cell.type_2_plot) {
  print(i)
  ds = Seurat.object
  gl = 6
  if (i == 'MCR') {gl = 3}
  HP.col <- DGE_Heatmap(Seurat.object = ds, 
                        cell_type = i, 
                        enrichR.path = indir,
                        CND = 'up', 
                        database = c('Reactome_2016','GO_Biological_Process_2018'), 
                        pathway_to_plot[[i]],
                        gene.label = gl)
  
  pdf(paste(outdir,'HM_DS_',i,'.pdf',sep=''),  width=7, height=4.5)
  print(HP.col)
  dev.off()
}

######## Dotplots ##########à

pathways2plot = pathway_to_plot
pathways.dataframe = data.frame(cellType = character(),
                                Pathway = character(),
                                gene.ratio = numeric(),
                                p.value = numeric(),
                                p.value.adj = numeric(),
                                stringsAsFactors=FALSE)

all.cell.types= cell.type_2_plot

fx <- function(x) eval(parse(text=enrichR.table$Overlap[x]))
fx <- function(x) eval(parse(text=enrichR.table[x,]$Overlap))
database = c('Reactome_2016','GO_Biological_Process_2018')
L = list(LEY = 18, MYD = 6)
for (cell_type in all.cell.types) {
  print(cell_type)
  enrichR.file = paste(indir,"enrichR_DGE_",cell_type,'_',CND,'.xlsx',sep='')
  enrichR.table = data.frame()
  for (dat in database) {
    Table <- read.xlsx(xlsxFile = enrichR.file, 
                       sheet = dat, 
                       startRow = 1, 
                       colNames = TRUE,
                       rowNames = TRUE, 
                       detectDates = FALSE, 
                       skipEmptyRows = TRUE,
                       skipEmptyCols = TRUE,
                       na.strings = "NA", 
                       fillMergedCells = FALSE)
    
    enrichR.table = rbind(enrichR.table , Table)
  }
  enrichR.table <- enrichR.table[order(enrichR.table$Adjusted.P.value, decreasing = F),]
  N = length(pathways2plot[[cell_type]])
  pathways.dataframe.entry = data.frame(cellType = rep(cell_type, N),
                                        Pathway = pathways2plot[[cell_type]],
                                        gene.ratio = sapply(pathways2plot[[cell_type]], fx),
                                        p.value = enrichR.table[pathways2plot[[cell_type]],]$P.value,
                                        p.value.adj = enrichR.table[pathways2plot[[cell_type]],]$Adjusted.P.value)
  
  #pathways.dataframe <- rbind(pathways.dataframe, pathways.dataframe.entry)
  pathways.dataframe <- pathways.dataframe.entry[order(pathways.dataframe.entry$p.value, decreasing = F),]
  pathways.dataframe$Pathway.num = dim(pathways.dataframe)[1]:1
  pathways.dataframe$Pathway.num = as.factor(pathways.dataframe$Pathway.num )
  removeHomo <- function(str, stopwords='Homo') {
    x <- unlist(strsplit(as.character(str), " "))
    paste(x[!x %in% stopwords], collapse = " ")
  }
  removeSapiens <- function(str, stopwords='sapiens') {
    x <- unlist(strsplit(as.character(str), " "))
    paste(x[!x %in% stopwords], collapse = " ")
  }
  pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeHomo)
  pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeSapiens)
  pathways.dataframe$Pathway <- str_pad(pathways.dataframe$Pathway, width = 70, side = "right", pad = " ")
  
  
  PP = ggplot(pathways.dataframe, aes(Pathway.num,-log10(p.value))) + 
    geom_point(aes(size = gene.ratio), color = colors_Alf[cell_type,]$cols) +
    scale_size_continuous(range = c(5,15), name = "Gene ratio") +
    ylim(c(0,L[[cell_type]])) +
    coord_flip() +
    scale_x_discrete(breaks=pathways.dataframe$Pathway.num, 
                     labels=pathways.dataframe$Pathway,
                     position = "top") +
    theme(plot.title = element_text(color="black", size=18, face="bold.italic"),
          plot.subtitle = element_text(color="black", size=16, face="italic"),
          axis.text.x = element_text(angle = 90, face = "bold", color = 'black', size=16, hjust =1, family = "mono"), 
          axis.title.x = element_text(face = "bold", color = "black", size = 14),
          axis.text.y = element_text(angle = 0, face = "bold", color = 'black', size=16, family = "mono"),
          axis.title.y = element_text(face = "bold", color = "black", size = 14),
          legend.text = element_text(color = "black", size = 12),
          legend.title = element_text(face = "bold", color = "black", size = 14),
          panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
    labs(title = 'Upregulated Pathways in Reactome and GO Biological Process', 
         x = "Pathways", 
         y = "-Log10(pvalue)") 
  
  assign(paste('PP',cell_type,sep=''),PP)
  
  pdf(paste(outdir,'PP_',cell_type,'.pdf',sep=''),  width=15, height=4.5)
  print(PP)
  dev.off()    
}


for (cell_type in all.cell.types) {
  HM <- image_read_pdf(paste(outdir,"HM_DS_",cell_type,".pdf", sep=''),density = 160)
  PP <- image_read_pdf(paste(outdir,"PP_",cell_type,".pdf", sep=''), density = 160)
  im = image_append(c(HM, PP))
  image_write(im, path = paste(outdir,"HM_PP_",cell_type,".pdf", sep=''), format = "pdf")
  image_write(im, path = paste(outdir,"HM_PP_",cell_type,".tiff", sep=''), format = "tiff")
}


######### Downregulated pathways #######
Seurat.object = final_int_obj

Seurat.object$cell_type = Seurat.object$celltype
Idents(Seurat.object) <- "celltype.cond"
DefaultAssay(Seurat.object) <- "RNA"

CND = 'down'
#cell_types= c("LEY", "LEY2", "MYD", "SRT","END", "STRO") # "MCR"
cell_types= c("LEY", "MYD")

database = c('Reactome_2016','GO_Biological_Process_2018')
gene.l = list()
path.l = list()
gene.list = character()
indir = paste(dir, "enrichR/", sep='')
outdir = paste(dir, "Pathways/", sep='')
for (cell_type in cell_types) {
  print(cell_type)
  gene.list = character()
  enrichR.file = paste(indir,'enrichR_DGE_',cell_type,'_',CND,'.xlsx', sep='')
  
  
  enrichR.table = data.frame()
  for (dat in database) {
    Table <- read.xlsx(xlsxFile = enrichR.file, 
                       sheet = dat, 
                       startRow = 1, 
                       colNames = TRUE,
                       rowNames = TRUE, 
                       detectDates = FALSE, 
                       skipEmptyRows = TRUE,
                       skipEmptyCols = TRUE,
                       na.strings = "NA", 
                       fillMergedCells = FALSE)
    
    enrichR.table = rbind(enrichR.table , Table)
  }
  p = row.names(enrichR.table[enrichR.table$Adjusted.P.value < 0.05,])
  path.l[[cell_type]] = p
  
  for (pathway in p) {
    gene.list.tmp <- unlist(strsplit(enrichR.table[pathway,]$Genes, ';'))
    gene.list = c(gene.list, gene.list.tmp)
    
  }
  gene.l[[cell_type]] = unique(gene.list)
  write.table(gene.list, paste(outdir,'DW_',cell_type,'.txt',sep=''),quote=T, row.names=F, col.names=F)
}

prova2 = intersect(gene.l$LEY,gene.l$MYD)
length(prova2)

order.2.plot = c('LEY_iNOA', 'LEY_OA', 'MYD_iNOA','MYD_OA','SRT_iNOA','SRT_OA', 'UND_iNOA', 'UND_OA')
Seurat.object = subset(Seurat.object, idents = order.2.plot)

levels(Seurat.object) <- order.2.plot
DefaultAssay(Seurat.object) <- "RNA"

#markers.to.plot <- AgeGenesClustersTable_dw$gene.name[AgeGenesClustersTable_dw$cluster==3]
dp_down = DotPlot(Seurat.object, 
                  col.min = -1,
                  col.max = 1,
                  features = prova2, 
                  cols = c("lightblue", "red"), 
                  dot.scale = 10) +
  scale_size_continuous(range = c(0.2,5)) +
  RotatedAxis() +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, face = "bold", color = "black", size=15,vjust =1.1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 20),
        axis.text.y = element_text(angle = 0, color = "black", size=14, family = 'mono'), 
        axis.title.y = element_text(face = "bold", color = "black", size = 20),
        legend.text = element_text(face = "bold", color = "black", size = 16),
        legend.position="right",
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "Genes", y = "Cell type")

dp_down
pdf(paste(outdir,'DW_genes_exp.pdf',sep=''),width=8, height=10)
plot(dp_down)
dev.off()


p = rev(as.character(sort(Reduce(intersect, path.l))))
p

pathways.dataframe = data.frame(cellType = character(),
                                Pathway = character(),
                                gene.ratio = numeric(),
                                p.value = numeric(),
                                p.value.adj = numeric(),
                                stringsAsFactors=FALSE)

all.cell.types= cell_types
database = c('Reactome_2016','GO_Biological_Process_2018')
N=length(p)
CND = 'DOWN'
fx <- function(x) eval(parse(text=enrichR.table$Overlap[x]))
fx <- function(x) eval(parse(text=enrichR.table[x,]$Overlap))

for (cell_type in all.cell.types) {
  print(cell_type)
  enrichR.file = enrichR.file = paste(indir,'enrichR_DGE_',cell_type,'_',CND,'.xlsx', sep='')
  enrichR.table = data.frame()
  for (dat in database) {
    Table <- read.xlsx(xlsxFile = enrichR.file, 
                       sheet = dat, 
                       startRow = 1, 
                       colNames = TRUE,
                       rowNames = TRUE, 
                       detectDates = FALSE, 
                       skipEmptyRows = TRUE,
                       skipEmptyCols = TRUE,
                       na.strings = "NA", 
                       fillMergedCells = FALSE)
    
    enrichR.table = rbind(enrichR.table , Table)
  }
  
  pathways.dataframe.entry = data.frame(cellType = rep(cell_type, length(p)),
                                        Pathway = p,
                                        gene.ratio = sapply(p, fx),
                                        p.value = enrichR.table[p,]$P.value,
                                        p.value.adj = enrichR.table[p,]$Adjusted.P.value)
  pathways.dataframe <- rbind(pathways.dataframe, pathways.dataframe.entry)       
}

removeHomo <- function(str, stopwords='Homo') {
  x <- unlist(strsplit(as.character(str), " "))
  paste(x[!x %in% stopwords], collapse = " ")
}
removeSapiens <- function(str, stopwords='sapiens') {
  x <- unlist(strsplit(as.character(str), " "))
  paste(x[!x %in% stopwords], collapse = " ")
}
pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeHomo)
pathways.dataframe$Pathway <- sapply(pathways.dataframe$Pathway, removeSapiens)


pathways.dataframe$Pathway <- factor(x = pathways.dataframe$Pathway, 
                                     levels = sort(unique(pathways.dataframe$Pathway),decreasing=T))


coloriamo = colors_Alf[all.cell.types,]$cols

patplot = ggplot(data=pathways.dataframe, aes(x=Pathway, y=-log10(p.value), fill=cellType)) +
  geom_bar(stat="identity", position=position_dodge()) +
  scale_fill_manual(values=as.character(coloriamo)) +
  coord_flip() +
  scale_x_discrete(position = 'top') + 
  theme(plot.title = element_text(color="black", size=22, face="bold.italic"),
        plot.subtitle = element_text(color="black", size=16, face="italic"),
        axis.text.x = element_text(angle = 0, face = "bold", color = 'black', size=14, hjust =1), 
        axis.title.x = element_text(face = "bold", color = "black", size = 18),
        axis.text.y = element_text(angle = 0, face = "bold", size=14, family = 'mono'),
        axis.title.y = element_text(face = "bold", color = "black", size = 18),
        legend.text = element_text(face = "bold", color = "black", size = 14),
        panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid"))

pdf(paste(outdir,'DW_pathPLOT.pdf',sep=''),width=12, height=9)
plot(patplot)
dev.off()

patplot

layout <- '
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
AAAAAAABBBB
'

p1 = wrap_plots(B = patplot, A = dp_down, design = layout) + plot_layout(guides = 'collect')

pdf(paste(outdir,'Figure7_main.pdf',sep=''),width=20, height=9.5)
p1
dev.off()

im = image_read_pdf(paste(outdir,"Figure7_main.pdf",sep=''),density = 140)
image_write(im, path = paste(outdir,"Figure7_main.pdf.tiff", sep=''), format = "tiff")

col = as.character(colors_Alf[names(path.l),]$cols)

venn::venn(path.l, simplify=TRUE, opacity = 0.3, box = FALSE, ilab=TRUE, zcolor = col, ilcs = 1.5, sncs = 2)

pdf(paste(outdir,'Figure7_venn.pdf',sep=''),width=7, height=6.5)
print(venn::venn(path.l, simplify=TRUE, opacity = 0.3, box = FALSE, elipse = T, ilab=TRUE, zcolor = col, ilcs = 1.5, sncs = 2.5))
dev.off()
im = image_read_pdf(paste(outdir,"Figure7_venn.pdf",sep=''),density = 140)
image_write(im, path = paste(outdir,"Figure7_venn.pdf.tiff", sep=''), format = "tiff")

```




 SC Transform integration
```{r SC integration: iNOA vs CTL - 3 dataset}
############### SC integration ###############
somatic.list

#3000*1024^2 
options(future.globals.maxSize = 3145728000) 

start_time <- Sys.time()

for (i in 1:length(somatic.list)) {
    somatic.list[[i]] <- SCTransform(somatic.list[[i]], verbose = FALSE)
}

# prepare integration
somatic.features <- SelectIntegrationFeatures(object.list = somatic.list, nfeatures = 3000)
somatic.list <- PrepSCTIntegration(object.list = somatic.list, anchor.features = somatic.features, verbose = FALSE)
# anchors
somatic.anchors <- FindIntegrationAnchors(object.list = somatic.list, normalization.method = "SCT", anchor.features = somatic.features, verbose = FALSE)
# SCT integration
somatic.SCintegrated <- IntegrateData(anchorset = somatic.anchors, normalization.method = "SCT", verbose = FALSE)

somatic.SCintegrated <- RunPCA(somatic.SCintegrated, npcs = 30, verbose = FALSE)
ElbowPlot(somatic.SCintegrated)

somatic.SCintegrated <- RunUMAP(somatic.SCintegrated, reduction = "pca", dims = 1:20, verbose = FALSE)

DimPlot(somatic.SCintegrated, reduction = "umap", pt.size = 2)
DimPlot(somatic.SCintegrated, reduction = "umap", pt.size = 1, split.by = 'source', group.by = 'source')


# Visualization and clustering on the integrated dataset
res = 0.2
nPC = 20
somatic.SCintegrated <- FindNeighbors(somatic.SCintegrated, dims = 1:nPC, verbose = FALSE)
somatic.SCintegrated <- FindClusters(somatic.SCintegrated, resolution = res, verbose = FALSE)
somatic.SCintegrated <- RunUMAP(somatic.SCintegrated, dims = 1:nPC, verbose = FALSE)
somatic.SCintegrated <- RunTSNE(somatic.SCintegrated, dims = 1:nPC, verbose = FALSE)

p1 = DimPlot(somatic.SCintegrated, reduction = "umap", label = T, pt.size = 1) + 
#scale_color_manual(values = (col)) +
theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
      axis.title.x = element_text(face = "bold", color = "black", size = 24),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
      axis.title.y = element_text(face = "bold", color = "black", size = 24),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position="top",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
labs(subtitle = paste('Res = ', res,', nPC = ',nPC, sep = ''), x = "UMAP 1", y = "UMAP 2") + NoLegend()

p2 = DimPlot(somatic.SCintegrated, reduction = "umap", group.by = 'condition',label = F, pt.size = 0.5, order = TRUE) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
      axis.title.x = element_text(face = "bold", color = "black", size = 24),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
      axis.title.y = element_text(face = "bold", color = "black", size = 24),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position="top",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(subtitle = paste('Res = ', res,', nPC = ',nPC, sep = ''), x = "UMAP 1", y = "UMAP 2") 

p1 | p2


bys = DimPlot(somatic.SCintegrated, reduction = "umap", label = T, pt.size = 1, split.by = 'source', order = TRUE) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
      axis.title.x = element_text(face = "bold", color = "black", size = 24),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
      axis.title.y = element_text(face = "bold", color = "black", size = 24),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position="top",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(subtitle = paste('Res = ', res,', nPC = ',nPC, sep = ''), x = "UMAP 1", y = "UMAP 2") + NoLegend()
bys 


DimPlot(somatic.SCintegrated, split.by = 'condition', label = T)

end_time <- Sys.time()
end_time - start_time

head(somatic.SCintegrated[[]])

```

```{r SCintegration - marker genes and plots: iNOA vs CTL- 3 datasets}
#500*1024^2 
options(future.globals.maxSize = 524288000) 

dataset = 'GSE142585_GSE124263_GSE112013'
dataset_iNOA_dir = '/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/adultCTL_iNOA_SCintegration/'
dir.create(dataset_iNOA_dir, recursive = TRUE)

bys = DimPlot(somatic.SCintegrated, reduction = "umap", label = T, pt.size = 1, split.by = 'source', order = TRUE) + 
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
      axis.title.x = element_text(face = "bold", color = "black", size = 24),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
      axis.title.y = element_text(face = "bold", color = "black", size = 24),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position="top",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(subtitle = paste('Res = ', res,', nPC = ',nPC, sep = ''), x = "UMAP 1", y = "UMAP 2") + NoLegend()

png(paste(dataset_iNOA_dir,"umapbysources.UPD_allcell_integration_",dataset,"_iNOA_res",res,".png",sep=''),  width = 1600, height = 500)
plot(bys)
dev.off()

seeMG = TRUE

if (seeMG) {
  filename_xlsx = paste(dataset_iNOA_dir, 'integration_',dataset,'_iNOA_res',res,'_2.xlsx', sep ='')
  cluster.markers = FindAllMarkers(somatic.SCintegrated)
  top10 <- cluster.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
  write.xlsx(cluster.markers,
            file= filename_xlsx, 
               row.names = T,
               asTable = T)
  HM =  DoHeatmap(somatic.SCintegrated, features = top10$gene, angle = 90) + NoLegend() + 
    scale_fill_gradientn(colours = coolwarm(200)) 
  png(paste(dataset_iNOA_dir,"somatic.HM.UPD_allcell_integration_",dataset,"_iNOA_res",res,".png",sep=''),  width = 1400, height = 1400)
  plot(HM)
  dev.off()
}

write.csv(cluster.markers, paste(dataset_iNOA_dir, 'seurat_MG_integration_',dataset,'_iNOA_res',res,'.csv', sep =''))

Seurat.object = somatic.SCintegrated

leydig_signature = c('CFD','DLK1','LUM','CALB2')
myoid_signature = c('ACTA2','MYH11','DES', 'MYL9')
sertoli_signature = c('FATE1','CITED1','SOX9','AMH','CLDN11')
macrophage_signature = c('CD14','CD74','HLA-DRA','HLA-DRB1')
endothelial_signature = c('VWF', "EGFL7",'CD34', 'PRSS23', "RBP7")
t_cell_signature =  c('GZMA','GZMK','CD2','CCL5','NKG7')

DefaultAssay(Seurat.object) <- 'RNA'
pL <- Plot_sign(Seurat.object,
          signature= c('CFD','DLK1','LUM'), 
          operator = mean, title = 'LEY')
pM <- Plot_sign(Seurat.object,
          signature= c('ACTA2','MYH11','DES'), 
          operator = mean, title = 'MYD')
pS <- Plot_sign(Seurat.object,
          signature= c('FATE1','CITED1','SOX9', 'AMH'), 
          operator = mean, title = 'SRT')
pMa <- Plot_sign(Seurat.object,
          signature= c('CD14','CD74','HLA-DRA'), 
          operator = mean, title = 'MCR')
pE <- Plot_sign(Seurat.object,
          signature= c('VWF',"EGFL7", 'PRSS23'), 
          operator = mean, title = 'END')
pT <- Plot_sign(Seurat.object,
          signature= c('GZMA','CD2','CCL5'), 
          operator = mean, title = 'TCL')
pSTRO <- Plot_sign(Seurat.object,
          signature= c('RGS5','TPM2','IGFBP5'), 
          operator = mean, title = 'STRO')
layout <- '
AABBCC##
AABBCC##
AABBCCDD
EEFFGGDD
EEFFGG##
EEFFGG##
'
 
wrap_plots(A = pL, B = pM, C = pS, G = pT, E = pE, F = pMa, D =pSTRO, design = layout)

pdf(paste(dataset_iNOA_dir, "MG_integration_",dataset,"_iNOA_res",res,".pdf",sep=''),  width=14, height=6)
wrap_plots(A = pL, B = pM, C = pS, D = pSTRO, E = pE, F = pMa, G =pT, design = layout)
dev.off()
png(paste(dataset_iNOA_dir, "MG_integration_",dataset,"_iNOA_res",res,".png",sep=''),  width=1400, height=600)
wrap_plots(A = pL, B = pM, C = pS, D = pSTRO, E = pE, F = pMa, G =pT, design = layout)
dev.off()


```


```{r SC cluster assignment: iNOA vs CTL - 3 dataset}
new.cluster.ids <- c( "LEY",
                      "MYD", 
                      "END",
                      "MCR",
                      "STRO",
                      "SRT",
                      "TCL")

names(new.cluster.ids) <- levels(somatic.SCintegrated)
Adult.somatic.SCintegrated <- RenameIdents(somatic.SCintegrated, new.cluster.ids)
SO = Adult.somatic.SCintegrated

col = as.character(colors_Alf[levels(SO),]$cols)

png(paste(dataset_iNOA_dir,"UMAP_cellAssigned_",dataset,"_iNOA_res",res,".png",sep=''),  width = 500, height = 400)
DimPlot(SO, reduction = "umap", label = TRUE, pt.size = 1, order = TRUE) +
  scale_color_manual(values = (col)) +
  theme(plot.title = element_text(color="black", size=26, face="bold.italic"),
      axis.text.x = element_text(angle = 90, face = "bold", color = "black", size=22, hjust =1), 
      axis.title.x = element_text(face = "bold", color = "black", size = 24),
      axis.text.y = element_text(angle = 0, face = "bold", color = "black", size=22),
      axis.title.y = element_text(face = "bold", color = "black", size = 24),
      legend.text = element_text(face = "bold", color = "black", size = 12),
      legend.position="right",
      panel.background = element_rect(fill = "white",colour = "black", size = 1, linetype = "solid")) +
  labs(x = "UMAP 1", y = "UMAP 2") 
dev.off()

MGnewname = TRUE
if (MGnewname) {
  filename_xlsx = paste(dataset_iNOA_dir, 'integration_',dataset,'_iNOA_res',res,'_celltype.xlsx', sep ='')
  cluster.markers = FindAllMarkers(SO)
  top10 <- cluster.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
  write.xlsx(cluster.markers,
              file= filename_xlsx, 
              row.names = T,
              asTable = T)
  HM =  DoHeatmap(SO, features = top10$gene, angle = 90) + NoLegend() + 
    scale_fill_gradientn(colours = coolwarm(200)) 
  
  png(paste(dataset_iNOA_dir,"somatic.HM.UPD_allcell_integration_",dataset,"_iNOA_res",res,".png",sep=''),  width = 1400, height = 1400)
  plot(HM)
  dev.off()
}
```

```{r SC - DGE and enrichment: iNOA vs CTL - 3 dataset}

######## DGE ###########
SO = Adult.somatic.SCintegrated
dir_DGE = paste(dataset_iNOA_dir,'DGE/', sep='')
dir.create(dir_DGE)

DefaultAssay(SO) <- "RNA"
# relabel cells to perform DGE
SO$celltype.cond <- paste(Idents(SO), SO$condition, sep = "_")
SO$celltype <- Idents(SO)
Idents(SO) <- "celltype.cond"
DGEresults = lapply(new.cluster.ids, myDGE, Seurat.object = SO, outdir = dir_DGE)

######## enrichment ###########
suppressMessages(library(enrichR))
dir_enrichR = paste(dataset_iNOA_dir,'enrichR/', sep='')
dir.create(dir_enrichR)

databases <- listEnrichrDbs()
# databases to make the enrichment of
enrich.databases <- c("GO_Biological_Process_2018",
                      "GO_Cellular_Component_2018",
                      "GO_Molecular_Function_2018",
                      "Reactome_2016",
                      "KEGG_2016",
                      "WikiPathways_2016",
                      "BioCarta_2016",
                      "BioPlanet_2019")


cell_types = new.cluster.ids
DGE_gene_up_list <- list()
DGE_gene_dw_list <- list()

for (celltype in cell_types) {
  print(celltype)
  DGE_file = paste(dir_DGE,"DGE_", celltype, ".iNOA.vs.CTRL.xlsx", sep= '')
  A <- read.xlsx(DGE_file)
  
  up.genes <- A$row.names[A$avg_logFC>0 & A$p_val_adj <= 0.05]
  down.genes <- A$row.names[A$avg_logFC<0 & A$p_val_adj <= 0.05]
  both.genes <- A$row.names[A$p_val_adj <= 0.05]
  
  DGE_gene_up_list[[celltype]] <- up.genes
  DGE_gene_dw_list[[celltype]] <- down.genes
  
  enrichr.list <- list()
  enrichr.list <- lapply(list(up.genes,down.genes,both.genes), function(x) {
    enrichR::enrichr(genes = x, databases = enrich.databases)
  })
  
  names(enrichr.list) <-  c("up","down","both")
  
  for (i in names(enrichr.list)){
    for (dat in names(enrichr.list[[i]])) {
      enrichr.list[[i]][[dat]] = enrichr.list[[i]][[dat]][,
                                                          c('Term', 
                                                            'Overlap', 
                                                            'P.value', 
                                                            'Adjusted.P.value', 
                                                            'Odds.Ratio', 
                                                            'Combined.Score', 
                                                            'Genes')]
    }
  }
  for (i in 1:length(enrichr.list)) {
    filename = paste(dir_enrichR,'enrichR_DGE_',celltype,'_',names(enrichr.list)[i],'.xlsx', sep='')
    if (!is.null(enrichr.list[[i]])) {write.xlsx(x = enrichr.list[[i]], file = filename, asTable = T)}
  }
}

venn::venn(DGE_gene_up_list, zcolor = rainbow(7))
venn::venn(DGE_gene_dw_list, zcolor = rainbow(7))
```




```{r save objects}
# save integrated objects
SO_dir = "/Users/tascini.annasofia/OneDrive - Ospedale San Raffaele/prj_COSR/AlfanoM_904_infertilita_epigenetics/Seurat_objects/"
dir.create(SO_dir)
save(GSE124263.somatic.integrated, file = paste(SO_dir,'GSE124263.somatic.integrated', sep =''))
save(GSE142585.somatic.integrated, file = paste(SO_dir,'GSE142585.somatic.integrated', sep =''))
save(Adult.somatic.integrated, file = paste(SO_dir,'Adult.somatic.integrated', sep =''))
save(somatic.SCintegrated, file = paste(SO_dir, "somatic.SCintegrated", sep = ''))
```


**Integration of testis somatic cells**
Sc Transform 
```{r allTogether}
allTogetherNow = FALSE
if (allTogetherNow == TRUE) {
  
testis.list = list(iNOA_D1 = iNOA_donor1,
                   iNOA_D2 = iNOA_donor2,
                   iNOA_D2 = iNOA_donor3,
                   GSE112013 = testis_somatic_GSE112013, # Guo adult
                   GSE120506 = testis_somatic_GSE120506, # Guo infant
                   GSE142585 = testis_somatic_GSE142585, # Shami et al
                   GSE106487 = testis_somatic_GSE106487, # Wang at al
                   GSE109037 = testis_somatic_GSE109037, # Hermann et al
                   GSE124263 = testis_somatic_GSE124263, # Sohni et al
                   GSE134144 = testis_somatic_GSE134144, # pubertal Guo et al 2020
                   GSE134144_t = testis_somatic_GSE134144_t, # transfemale Guo et al 2020
                   GSE125372 = testis_somatic_GSE125372) # Xia et al.

for (i in 1:length(testis.list)) {
    testis.list[[i]] <- SCTransform(testis.list[[i]], verbose = FALSE)
}

# prepare integration
testis.features <- SelectIntegrationFeatures(object.list = testis.list, nfeatures = 3000)
testis.list <- PrepSCTIntegration(object.list = testis.list, anchor.features = testis.features, verbose = FALSE)
# anchors
testis.anchors <- FindIntegrationAnchors(object.list = testis.list, normalization.method = "SCT", anchor.features = testis.features, verbose = FALSE)
# SCT integration
testis.integrated <- IntegrateData(anchorset = testis.anchors, normalization.method = "SCT", verbose = FALSE)
# Visualization and clustering on the integrated dataset
testis.integrated[[]]
testis.integrated <- RunPCA(testis.integrated, verbose = FALSE)
testis.integrated <- RunUMAP(testis.integrated, dims = 1:30)
plots <- DimPlot(testis.integrated, group.by = c("tech", "celltype"))
}
```











